# Scripts used in the project

This repository hosts the scripts and instructions to obtain the results presented in Cruz DÃ¡valos et al. (2022) [https://doi.org/10.1101/2022.01.27.477466].

Some of the analyses include:
* Visualization of geographic regions
* Mapping NGS reads (which was developed in parallel with the [mapache](https://github.com/sneuensc/mapache/) repository)
* Metagenomic screening
* Phylogenetic analyses
* Assessment data quality
* Determine genomic sex
* Infer molecular damage
* Estimate levels of contamination
* Isotopic analyses (diet reconstruction and calibration of radiocarbon dates)
* Dimension reduction (e.g., PCA, MDS)
* Genotype likelihood calculation and genotype calling
* Admixture analyses
* Assessment of relatedness between individuals
* Population genetics analyses with admixtools
* Inference of runs of homozygosity
* Demographic inference
* Data visualization (in general)

Please see [Figures_Paper.html](https://github.com/dianaicd/GenomicsMetagenomics_IndigenousBrazil/blob/master/Figures_Paper.html) and [Fgures_Paper.Rmd](https://github.com/dianaicd/GenomicsMetagenomics_IndigenousBrazil/blob/master/Figures_Paper.Rmd) for the code to make the figures that made it to the final manuscript.
