---
title: "Report_2017_12_14"
author: "Cruz-Dávalos, Diana I."
date: "12/11/2017"
output: html_document
---
---
title: "Botocudos project"
author: "Cruz-Dávalos, Diana I."
date: "9/19/2017"
output: 
  html_document:
    toc: True
    toc_float:
      collapsed: False
      smooth_scroll: True
    number_sections: False
    theme: "lumen"
    highlight: "tango"
    df_print: paged

---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = T)
```

```{r sources, echo = F, message=F, warning=F, cache = F}
source("~/Projects/Botocudos/Scripts/translate_ids.R")
source("~/Projects/Botocudos/Scripts/length_distribution_plot.R")
source("~/Projects/Botocudos/Scripts/preseq_only_plots.R")
source("~/Projects/Botocudos/Scripts/ry_plot.R")
source("~/Projects/Botocudos/Scripts/engoenous_plot.R")
source("~/Projects/Botocudos/Scripts/mapDamage_plot.R")
source("~/Projects/Botocudos/Scripts/reads_summaries_plots.R")
source("~/Projects/Botocudos/Scripts/order_endo.R")
source("~/Projects/Botocudos/Scripts/plots_coverage.R")
source("~/Projects/Botocudos/Scripts/pca_plot.R")
source("~/Projects/Botocudos/Scripts/pct_var_pc.R")
source("~/Projects/Botocudos/Scripts/plot_bammds.R")
```

```{r, echo = F}
library(cowplot)
library(ggplot2)
library(adegenet)
library(plyr)
#library(shiny)
boto <- read.table("~/Projects/Botocudos/Files/Summaries/Botocudos_summary_2017_09_18.table", header = T)
new_boto <- read.table("~/Projects/Botocudos/Files/Summaries/Botocudos_summary_2017_12_06.table", header = T)

boto$Library <- factor(boto$Library, levels = 
                         c("*",as.vector(order_endo( 0, 0.5, boto)$Library))
                       ) 
new_boto$Library <- factor(boto$Library, levels = 
                         c("*",as.vector(order_endo( 0, 0.5, new_boto)$Library))
                       ) 
boto$seq_round <- "First stage"
new_boto$seq_round <- "Second stage"
```


# Samples

We analyzed a total of 23 individuals: one labeled as "Native American" (~600 years old\*) from Minas Gerais, one identified as "Shell mound" (~4,000 years old\*) from Santa Catarina, and twenty-one "Botocudos". From this last group, eight skulls come from Espírito Santo, six from Minas Gerais, one from Bahía, and the 6 remaining have unassigned procedence (probably Minas Gerais or Espírito Santo).

\*according to the Museum's Excell file. We are currently dating the samples.

```{r, fig.align="left", echo = F, warning=F, message=F}
library(knitr)
include_graphics("Brazil_map_points.png", dpi = 120)
```

*Red: Minas Gerais

*Green: Espírito Santo

*Yellow: Bahía

*Pink: Santa Catarina

```{r boto table, echo = F}
library(DT)

datatable(new_boto[new_boto$Library!="*",], 
          filter = "bottom", 
          selection = "multiple",
          fillContainer = T, 
          rownames = F,
          extensions = 
            'FixedColumns',
          options = list(
              dom = 't',
              scrollX = TRUE,
              scrollY  = TRUE,
              fixedColumns = list(
                leftColumns = 3, rightColumns = 0))
          )
           # options = list(
           #   #oButtons = list(
           #     dom = 'Bfrtip',
           #     buttons = c('colvis','copy', 'csv', 
           #                 'excel', 'pdf', 'print')
           #  ))#,
           #  # oFixedColumns = list(
            #   dom = 't', 
            #   scrollX = F,
            #   scrollY  = F,
            #   fixedColumns = list(leftColumns = 3, 
            #                       rightColumns = 0)
            # ),
            # I("colvis")

              

```

# Preseq output

## Complexity curves


### <1%
```{r preseq expreads1, echo = F, fig.height=12, fig.width=12, fig.align="center"}
load("~/Projects/Botocudos/Files/PreSeq/output_preseq_Sep16_all.Rda")

### Breaks
# <1%
ind <- order_endo(Min = 0, Max = 0.01, boto)
plots <- vector("list", dim(ind)[1])
j <- 1
for(i in ind$Library){
  plots[[j]] <- plot_expreads_preseq(i, ylim = 2e+6, verify = T)
  j <- j+1
}
plot_grid(plotlist = plots, ncol = 3)


```

### 1% - 5%
```{r preseq expreads2, echo = F, fig.height=8, fig.width=12, fig.align="center"}
# 1 - 5%
ind <- order_endo(Min = 0.01, Max = 0.05, boto)
plots <- vector("list", dim(ind)[1])
j <- 1
for(i in ind$Library){
  plots[[j]] <- plot_expreads_preseq(i, ylim = 1e+7, verify = T)
  j <- j + 1
}
plot_grid(plotlist = plots, ncol = 3)
```

### 5% - 10%
```{r preseq expreads3, echo = F, fig.height=12, fig.width=12, fig.align="center"}
# 6 - 10%
ind <- order_endo(Min = 0.05, Max = 0.11, boto)
plots <- vector("list", dim(ind)[1])
j <- 1
for(i in ind$Library){
  plots[[j]] <- plot_expreads_preseq(i, ylim = 2e+7, verify = T)
  j <- j + 1
}
plot_grid(plotlist = plots, ncol = 3)

```

### 17% - 18%
```{r preseq expreads4, echo = F, fig.height=4, fig.width=8, fig.align="center"}
# 17-18%
ind <- order_endo(Min = 0.11, Max = 0.3, boto)
plots <- vector("list", dim(ind)[1])
j <- 1
for(i in ind$Library){
  plots[[j]] <- plot_expreads_preseq(i, ylim = 5e+7, verify = T)
  j <- j + 1
}

plot_grid(plotlist = plots, ncol = 2)
```

### 35%
```{r preseq expreads5, echo = F, fig.height=4, fig.width=4, fig.align="center"}

# 35%
ind <- order_endo(Min = 0.3, Max = 0.4, boto)
plots <- vector("list", dim(ind)[1])
j <- 1
for(i in ind$Library){
  plots[[j]] <- plot_expreads_preseq(i, ylim = 6e+7, verify = T)
  j <- j + 1
}

print(plots)
```


# Summary of sequenced reads

```{r summary plot, echo = F, fig.align='center', fig.width = 14}
p <- summary_reads_plot(boto)
p_lane <- summary_reads_plot(new_boto)
p_lane <- p_lane + geom_segment(aes(x = 0.5, xend=7.5, y=-1e6, yend=-1e6), 
                                color = "darkgrey") +
  geom_segment(aes(x = 7.5, xend=13.5, y=-1e6, yend=-1e6), color = "black") +
  geom_segment(aes(x = 13.5, xend=20, y=-1e6, yend=-1e6), 
               color = "darkgrey") +
  geom_segment(aes(x = 20, xend=23.5, y=-1e6, yend=-1e6), color = "black") +
  labs(y = NULL) +
  geom_label(label = "Lane 1", x = 4, y = -3.5e6,
             color = "black", fill = "white", label.size = NA) +
  geom_label(label = "Lane 2", x = 10, y = -3.5e6,
             color = "black", fill = "white", label.size = NA) +
  geom_label(label = "Lane 3", x = 17, y = -3.5e6,
             color = "black", fill = "white", label.size = NA) +
  geom_label(label = "Lane 4", x = 21.5, y = -3.5e6,
             color = "black", fill = "white", label.size = NA) +
  labs(y = NULL)

plot_grid(p, p_lane, rel_widths = c(4,4))



```

Reads aligned to the Human genome: 

After the second round of sequencing, we obtained between 14,811 and 19136109 reads mapping to the human genome per sample, with a depth of coverage up to 0.34x.

```{r aligning to hg, echo =F, fig.align='center', fig.width = 14}

p <- hg_mapping_plot(boto, ylim=2.2e7)
p_lane <- hg_mapping_plot(new_boto)
p_lane <- p_lane + geom_segment(aes(x = 0.5, xend=7.5, y=-1e5, yend=-1e5),
                                color = "darkgrey") +
  geom_segment(aes(x = 7.5, xend=13.5, y=-1e5, yend=-1e5),
               color = "black") +
  geom_segment(aes(x = 13.5, xend=20.5, y=-1e5, yend=-1e5),
               color = "darkgrey") +
  geom_segment(aes(x = 20, xend=23.5, y=-1e5, yend=-1e5),
               color = "black") +
  geom_label(label = "Lane 1", x = 4, y = -8e5,
             color = "black", fill = "white", label.size = NA) +
  geom_label(label = "Lane 2", x = 10, y = -8e5,
             color = "black", fill = "white", label.size = NA) +
  geom_label(label = "Lane 3", x = 17, y = -8e5,
             color = "black", fill = "white", label.size = NA) +
  geom_label(label = "Lane 4", x = 21.5, y = -8e5,
             color = "black", fill = "white", label.size = NA) +
  labs(y = NULL)

plot_grid(p, p_lane, rel_widths = c(4,4))
```

```{r endogenous, echo = F, fig.align='center', fig.width = 14}
#good <- sum(unique(boto$hits_unique_frac.endogenous.[boto$Library != "*"]) >0.01)

p <- ggendogenous(boto)
p_lane <- ggendogenous(new_boto)
p_lane <- ggendogenous(rbind(boto[boto$Library != "*",],
                             new_boto[new_boto$Library != "*",]), 
                       fill = "seq_round", color = c("firebrick2", "darkturquoise"),
                       lanes = T)

p_lane <- p_lane  +
  geom_label(label = "Pool 1", x = 4, y = -0.012,
             color = "black", fill = "white", label.size = NA) +
  geom_label(label = "Pool 2", x = 10, y = -0.012,
             color = "black", fill = "white", label.size = NA) +
  geom_label(label = "Pool 3", x = 17, y = -0.012,
             color = "black", fill = "white", label.size = NA) +
  geom_label(label = "Pool 4", x = 21.5, y = -0.012,
             color = "black", fill = "white", label.size = NA) +
  labs(y = NULL) + 
  geom_segment(aes(x = 0.5, xend=7.5, y=-0.005, yend=-0.005),
               color = "darkgrey") +
  geom_segment(aes(x = 7.5, xend=13.5, y=-0.005, yend=-0.005),
               color = "black") +
  geom_segment(aes(x = 13.5, xend=20.5, y=-0.005, yend=-0.005),
               color = "darkgrey") +
  geom_segment(aes(x = 20.5, xend=23.5, y=-0.005, yend=-0.005),
               color = "black")

pdf("~/Projects/Botocudos/Scripts/examples/Thesis Templates/RMarkdownTemp/figures/endogenous.pdf",
    width = 20, height = 6)
plot_grid(p, p_lane, rel_widths = c(4,4), labels = "AUTO")
dev.off()


png("~/Projects/Botocudos/Docs/Figures/endogenous.png",
    width = 10, height = 6, res = 200, units = "in")
p_lane + theme(legend.position = "top")
dev.off()

```

# Coverage

```{r bases covered, echo = F, fig.align="center", fig.width = 14}
p <- plot_bases_covered(boto)
p_lane <- plot_bases_covered(new_boto)

p_lane <- p_lane +
  geom_label(label = "Lane 1", x = 4, y = -3e7,
             color = "black", fill = "white", label.size = NA) +
  geom_label(label = "Lane 2", x = 10, y = -3e7,
             color = "black", fill = "white", label.size = NA) +
  geom_label(label = "Lane 3", x = 17, y = -3e7,
             color = "black", fill = "white", label.size = NA) +
  geom_label(label = "Lane 4", x = 21.5, y = -3e7,
             color = "black", fill = "white", label.size = NA) +
  labs(y = NULL) +
  geom_segment(aes(x = 0.5, xend=7.5, y=-1e7, yend=-1e7), 
               color = "darkgrey") +
  geom_segment(aes(x = 7.5, xend=13.5, y=-1e7, yend=-1e7), color = "black") +
  geom_segment(aes(x = 13.5, xend=20, y=-1e7, yend=-1e7), 
               color = "darkgrey") +
  geom_segment(aes(x = 20, xend=23.5, y=-1e7, yend=-1e7), color = "black") +
  labs(y = NULL) 
plot_grid(p, p_lane, rel_widths = c(4,4))

p <- plot_depth_cov(boto)
p_lane <- plot_depth_cov(new_boto)

p_lane <- p_lane  +
  geom_label(label = "Lane 1", x = 4, y = -0.012,
             color = "black", fill = "white", label.size = NA) +
  geom_label(label = "Lane 2", x = 10, y = -0.012,
             color = "black", fill = "white", label.size = NA) +
  geom_label(label = "Lane 3", x = 17, y = -0.012,
             color = "black", fill = "white", label.size = NA) +
  geom_label(label = "Lane 4", x = 21.5, y = -0.012,
             color = "black", fill = "white", label.size = NA) +
  labs(y = NULL) + 
  geom_segment(aes(x = 0.5, xend=7.5, y=-0.005, yend=-0.005),
               color = "darkgrey") +
  geom_segment(aes(x = 7.5, xend=13.5, y=-0.005, yend=-0.005),
               color = "black") +
  geom_segment(aes(x = 13.5, xend=20.5, y=-0.005, yend=-0.005),
               color = "darkgrey") +
  geom_segment(aes(x = 20, xend=23.5, y=-0.005, yend=-0.005),
               color = "black")

plot_grid(p, p_lane, rel_widths = c(4,4))
```


```{r, echo = F, fig.align='center'}


ggplot(rbind(boto[boto$Library != "*",], new_boto[boto$Library != "*",]), 
       aes(x = Library, y = hits_length.nuclear., fill = seq_round)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  theme(axis.text.x = element_text(angle = 90)) +
  labs(y = "Average length (bp)", title = "Length", x = NULL) +
  scale_fill_manual(values = c("darkturquoise", "darkviolet"),
                    name = "Sequencing") +
  geom_label(label = "Lane 1", x = 4, y = -2,
             color = "black", fill = "white", label.size = NA) +
  geom_label(label = "Lane 2", x = 10, y = -2,
             color = "black", fill = "white", label.size = NA) +
  geom_label(label = "Lane 3", x = 17, y = -2,
             color = "black", fill = "white", label.size = NA) +
  geom_label(label = "Lane 4", x = 21.5, y = -2,
             color = "black", fill = "white", label.size = NA) +
  labs(y = NULL) +
  geom_segment(aes(x = 0.5, xend=7.5, y=-1, yend=-1), 
               color = "darkgrey") +
  geom_segment(aes(x = 7.5, xend=13.5, y=-1, yend=-1), color = "black") +
  geom_segment(aes(x = 13.5, xend=20, y=-1, yend=-1), 
               color = "darkgrey") +
  geom_segment(aes(x = 20, xend=23.5, y=-1, yend=-1), color = "black") 

ggplot(rbind(boto[boto$Library != "*",], new_boto[boto$Library != "*",]), 
       aes(x = Library, y = hits_clonality.nuclear.,
           fill = seq_round)) +
  geom_bar(stat = "identity", position = "dodge") +
  theme(axis.text.x = element_text(angle = 90)) +
  labs(y = "Clonality (fraction)", title = "Clonality", x = NULL) +
  scale_fill_manual(values = c("darkturquoise", "darkviolet"),
                    name = "Sequencing") +
  geom_label(label = "Lane 1", x = 4, y = -0.002,
             color = "black", fill = "white", label.size = NA) +
  geom_label(label = "Lane 2", x = 10, y = -0.002,
             color = "black", fill = "white", label.size = NA) +
  geom_label(label = "Lane 3", x = 17, y = -0.002,
             color = "black", fill = "white", label.size = NA) +
  geom_label(label = "Lane 4", x = 21.5, y = -0.002,
             color = "black", fill = "white", label.size = NA) +
  labs(y = NULL) +
  geom_segment(aes(x = 0.5, xend=7.5, y=-0.001, yend=-0.001), 
               color = "darkgrey") +
  geom_segment(aes(x = 7.5, xend=13.5, y=-0.001, yend=-0.001), color = "black") +
  geom_segment(aes(x = 13.5, xend=20, y=-0.001, yend=-0.001), 
               color = "darkgrey") +
  geom_segment(aes(x = 20, xend=23.5, y=-0.001, yend=-0.001), color = "black") 

```

# Sex determination

```{r, echo = F, align = "center", fig.height = 8, fig.width = 12}
p <- ggsex(boto = boto)
p <- p + theme(legend.position = "none")
p_2 <- ggsex(boto = boto,
                d = "~/Projects/Botocudos/Files/Sex_determination/2017_12_06/")
p_merged <- ggsex(boto = boto,
                d = "~/Projects/Botocudos/Files/Sex_determination/2017_12_06/Merged/")

leyenda <- get_legend(p_merged)
p_2 <- p_2 + theme(legend.position = "none")
p_merged <- p_merged + theme(legend.position = "none")

png("~/Projects/Botocudos/Docs/Figures/sex.png", width =18, height = 8,
    units = "in", res = 200)
plot_grid(p, p_2, p_merged, leyenda, ncol = 4, 
          labels = c("A", "B", "C", NULL), rel_widths = c(2,2,2,2))
dev.off()

```

# NGSAdmix

```{r, echo = F, fig.width = 18}
setwd("~/Projects/Botocudos/Files/ADMIX/2017_12_06/")
k <- 2
sufix <- "America_Oceania_nodamage_8botocudos_k"
sufix <- "America_Oceania_nodamage_homozygous_8botocudos_k"
name <- paste(sufix, k, ".qopt", sep = "")
admix<-t(as.matrix(read.table(name)))
colors<- funky(k)
#barplot(admix, space=0,border=NA,xlab="Individuals",ylab="admixture",
#        col = colors)
panel_names <- "botos_america_oceania.popinfo"

other_dir <- "~/Projects/Botocudos/Files/ADMIX/2017_12_06/"
pop<-read.table(paste(other_dir,panel_names,
sep = ""), as.is=T)
pop_order <- factor(pop[,1], levels=c("Botocudos", "Karitiana", "Surui", "Chane", "Peruvian",
                                      "Quechua", "Colombians", "Mayan", "Mixtec", "Mixe", "Zapotec", "MXL", "Pima", "Hawaiian", "Igorot","Dusun",
                                      "NAN_Melanesian", "Bougainville",  "Papuan", "Australian", "Maori"))
pop <- pop[order(pop_order),]
png("~/Projects/Botocudos/Docs/Figures/ngsadmix_nodamage_homozygous.png",
    width = 17, height = 20, units = "in", res = 200)
par(mfrow = c(1, 10))
for(k in seq(2,10)){  
  par(mar = c(6, 0.5, 4, 2))
  name <- paste(sufix, k, ".qopt", sep = "")
  colors<- funky(k)
  admix<-t(as.matrix(read.table(name)))
  #admix<-admix[,order(pop[,1])]
  admix <- admix[, order(pop_order)]
  #pop<-read.table(paste(other_dir,panel_names, sep = ""), as.is=T)
  #pop <- pop[order(pop_order),]
  barplot(admix,col=colors,space=0,border=NA,
          ylab=NULL,
          xlab="Admixture\n proportion",
          main = paste("k =", k), horiz = T, cex.main = 3,
          cex.lab = 2)
  
  pop_names <- tapply(1:nrow(pop),pop[,1],mean)
  #text(-0.15, pop_names, names(pop_names),xpd=T,cex = 1.2)
  abline(h =  tapply(1:nrow(pop),pop[,1],max))
  #NGSadmixEx2.png 
  
}
text(1.8, pop_names, names(pop_names),xpd=NA,cex = 2)
dev.off()

```

# PCA

```{r, echo = F, fig.width = 18}
d <- "~/Projects/Botocudos/Files/PCA/2017_12_06/"
for(ind in unique(boto$Target)){
  p <- pca_plot(ind, plink=F, d, pcx = 1, pcy = 2, panel_name = "merged_noduplicates_reheaded")
  p <- p + theme(legend.position = "none")
  p2 <- pca_plot(ind, plink=F, d, pcx = 2, pcy = 3, panel_name = "merged_noduplicates_reheaded")
  x <- plot_grid(p, p2, ncol = 2, rel_widths = c(1, 1.2))
  print(x)
}

d <- "~/Projects/Botocudos/Files/PCA/2017_12_06/"
for(ind in unique(boto$Target)){
  p <- pca_plot(ind, plink=F, d, pcx = 1, pcy = 2, panel_name = "America_Oceania_nopapus_noaus")
  p <- p + theme(legend.position = "none")
  p2 <- pca_plot(ind, plink=F, d, pcx = 2, pcy = 3, panel_name = "America_Oceania_nopapus_noaus")
  x <- plot_grid(p, p2, ncol = 2, rel_widths = c(1,1.2))
  print(x)
}

selected <- c("MA2384", "MA2400", "MA2394", "MA2402", "MA2398",
              "MA2399", "MA2382", "MA2392")
snps <- c(2138, 2345, 1810, 4319, 9989, 11297, 15983, 31528)
names(snps) <- selected

for(ind in selected){
  p <- pca_plot(ind, plink=F, d, pcx = 1, pcy = 2, panel_name = "merged_noduplicates_reheaded")
  p <- ggdraw(p) + theme(legend.position = "none") +
    draw_label(paste(prettyNum(snps[ind], big.mark = ","),
                     "SNPs"),
               0.3, 0.9)
  p2 <- pca_plot(ind, plink=F, d, pcx = 2, pcy = 3, panel_name = "merged_noduplicates_reheaded")
  x <- plot_grid(p, p2, ncol = 2, rel_widths = c(1, 1.2))
  print(x)
}


```

# MDS

```{r, echo = F, fig.width=18}
d <- "~/Projects/Botocudos/Files/MDS/2017_12_06/"
ind <- "merged_noduplicates_reheaded_nodamage_8"

colors <- funky(10)
tmp <- colors[3]
colors[3] <- colors[6]
colors[6] <- tmp
colors <- c("black", colors)
plots <- plot_bammds(d = d, ind = ind, title = "All populations", cx = 1, cy = 2,
            colorby = "region", colors = colors)
plots2 <- plot_bammds(d = d, ind = ind, title = "All populations", cx = 2, cy = 3,
            colorby = "region", switch = T, colors = colors)

selected <- c( "MA2384", "MA2400",
 "MA2394", "MA2402", "MA2398", "MA2399",
 "MA2382", "MA2392")
selected <- rev(selected)
#snps <- c(191206, 125618, 78637, 70660, 38595, 22835, 22675, 25005)
snps <- c(36491, 23277, 15152, 13599, 7341, 4308, 4334, 4706)
plots[[1]] <- ggdraw(plots[[1]]) +
  draw_label(
    paste( sapply(selected, function(x) ma2mn(x)), 
           prettyNum(snps, big.mark = ","), sep = ": ",
           collapse = "\n"), 0.8, 0.8, size = 12)

png("~/Projects/Botocudos/Docs/Figures/mds_all_nodamage.png",
    width = 14, height = 5, res = 200, units = "in")
plot_grid(plots[[1]], plots2[[1]], plots2[[2]], ncol = 3, 
          rel_widths = c(3,3,1))
dev.off()

#ind <- "America_Oceania_reheaded_8"
ind <- "America_Oceania_nodamage_8"
plots <- plot_bammds(d = d, ind = ind, title = "America and Oceania", cx = 1, cy = 2,
            colorby = "population")
plots2 <- plot_bammds(d = d, ind = ind, title = "America and Oceania", cx = 2, cy = 3,
            colorby = "population", switch = T)

#snps <- c(191256, 125668, 78659, 70686, 38618, 22843, 22677, 25008)
snps <- c(36523, 23308, 15165, 13615, 7357, 4311, 4340, 4711)
plots[[1]] <- ggdraw(plots[[1]]) + draw_label(
  paste( sapply(selected, function(x) ma2mn(x)), 
         prettyNum(snps, big.mark = ","), sep = ": ",
         collapse = "\n"), 0.35, 0.2, size = 8)

png("~/Projects/Botocudos/Docs/Figures/mds_amoc_nodamage.png",
    width = 15, height = 6, res = 200, units = "in")
plot_grid(plots[[1]], plots2[[1]], plots2[[2]], ncol = 3, rel_widths = c(2,2,1))
dev.off()


ind <- "America_nodamage_8"

plots3 <- plot_bammds(d = d, ind = ind, title = "America", cx = 1, cy = 2,
            colorby = "population")
plots4 <- plot_bammds(d = d, ind = ind, title = "America", cx = 2, cy = 3,
            colorby = "population", switch = T)

#snps <- c(191263, 125668, 78670, 70686, 38620, 22846, 22681, 25011)
snps <- c(36537, 23329, 15177, 13625, 7360, 4316, 4341, 4712)
plots[[1]] <- ggdraw(plots[[1]]) + draw_label(
  paste( sapply(selected, function(x) ma2mn(x)), 
         prettyNum(snps, big.mark = ","), sep = ": ",
         collapse = "\n"), 0.8, 0.3, size = 8)

png("~/Projects/Botocudos/Docs/Figures/mds_america_nodamage.png",
    width = 14, height = 5, res = 200, units = "in")
plot_grid(plots[[1]], plots2[[1]], plots2[[2]], ncol = 3, rel_widths = c(3,3,1))
dev.off()


png("~/Projects/Botocudos/Docs/Figures/mds.png",
    width = 16, height = 10, res = 200, units = "in")
plot_grid(plots[[1]], plots2[[1]], plots2[[2]], 
          plots3[[1]], plots4[[1]], plots4[[2]],
          ncol = 3, rel_widths = c(3,3,1))
dev.off()

```

