---
title: "PCA_Viruses"
author: "Cruz-DÃ¡valos, Diana I."
date: "3/17/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, message = F, warning = F)
```

```{r}
library(cowplot)
library(factoextra)
library(plyr)
source("~/Projects/Botocudos/Scripts/misc/translate_ids.R")
```

The figure on top corresponds to a heatmap on the number of hits per virus,
divided by the total number of reads given to DIAMOND, and the figure below 
is a heatmap taking the logarithm on the same data.


```{r}
virus <- read.csv("~/Projects/Botocudos/Files/Microbiome/Virome/HumanVirus2.txt",
                  skip=2, header = T, sep = "\t")

rownames(virus) <- gsub("_", " ", gsub("-", " ", virus[,1]))

heatmap(as.matrix(virus[4:38,2:25]/as.numeric(virus[1, 2:25])), 
        col = rev(heat.colors(350)), 
        main = "Heatmap on normalized read counts")

heatmap(as.matrix(log(virus[4:38,2:25]/as.numeric(virus[1, 2:25])+ 1e-30)), 
        col = rev(heat.colors(350)), 
        main = "Heatmap on normalized read counts (log)")

```

The PCAs below were done on DIAMOND's results.
The number of hits per virus were divided by the amount
of starting within samples. 
We extracted the viruses that are reported to infect humans.
After these filters, a few viruses were removed as they might be false negatives,
e.g., Mamastrovirus (an RNA virus), Yaba monkey tumor virus 
(a virus found in monkeys in Nigeria, transmitted to humans working with monkeys).

The visualization of the data is done using the library `factoextra`.

We plotted the individuals on the first two dimensions, and the 10 variables (viruses) that contribute the most to the first two dimensions.

```{r}

data <- t(virus[4:38, 2:25]/as.numeric(virus[1, 2:25]))*1e6 + 1e-9
data <- data[, -which(colnames(data) == "Mamastrovirus 1")]
data <- data[, -grep("Yaba", colnames(data))]
data <- data[, -grep("retrovirus", colnames(data))]
```


```{r}
x <- prcomp((data))


p <- fviz_pca_ind(x, repel = T) 
p <- p + 
  geom_point(x = x$x[rownames(x$x) == "MN01701",1], 
             y = x$x[rownames(x$x) == "MN01701",2], col = "red") + 
  geom_point(x = x$x[rownames(x$x) == "MN1943",1], 
             y = x$x[rownames(x$x) == "MN1943",2], col = "blue")

y <- list(contrib = 10)
contrib <- fviz_pca_var(x, select.var = y, repel = T, col.var = "contrib")

var_exp <- fviz_screeplot(x)

```

```{r, fig.width=13, fig.height=8}
plot_grid(p, contrib, ncol = 2)
```

We observe that the Human alphaherpesvirus 3 (*i.e.*, varicella-zoster virus)
has a high contribution (about 40%) to the firs two components. Two other viruses are correlated with varicella, Human gammaherpesvirus 8 (agent causing Kaposi sarcoma) and NY 014 poxvirus (an unclassified poxvirus). As suggested by the heatmap, the first dimension is separating an individual with a high viral load from the rest of the population. The first two dimensions for this PCA explain 71.4% of the variance.

```{r}
print(var_exp)
```


For the following plots, we did a PCA using the log of the data, 
to reduce the "strength" of the contribution from some variables.

```{r}
x <- prcomp(log(data))


p <- fviz_pca_ind(x, repel = T) 
p <- p + 
  geom_point(x = x$x[rownames(x$x) == "MN01701",1], 
             y = x$x[rownames(x$x) == "MN01701",2], col = "red") + 
  geom_point(x = x$x[rownames(x$x) == "MN1943",1], 
             y = x$x[rownames(x$x) == "MN1943",2], col = "blue")

y <- list(contrib = 10)
contrib <- fviz_pca_var(x, select.var = y, repel = T, col.var = "contrib")

var_exp <- fviz_screeplot(x)
```

```{r, fig.width=13, fig.height=8}
plot_grid(p, contrib, ncol = 2)
```

Almost magically, Human Parvovirus B19 appears in the top 10 viruses.
We can say that as it is known to be a pretty small virus, we tried to reconstruct 
its genome with the data we had.

```{r}
print(var_exp)
```

```{r}

pv <- read.table("~/Projects/Botocudos/Files/Microbiome/Virome/Parvovirus/2019_03_14/geno2.txt")
colnames(pv)<- c("chr", "start", "end", "cov")
pv <- pv[pv$cov > 0,]
chrs <- unique(pv$chr)
l <- c(4354, 4767, 4922)
names(l) <- c("HQ340602.1", "DQ333427.1", "AJ717293.1")


par(mfcol=c(3,2))
for(i in seq(1, length(l))){
  plot(pv$start[pv$chr == names(l)[i]], pv$cov[pv$chr == names(l)[i]], 
     type = "n" , xlim = c(0, l[i]), bty = "n",
     main = paste("Genotype 2, strain", names(l)[i]),
     xlab = "Position (bp)", ylab = "Genome coverage")
segments(x0 =pv$start[pv$chr == names(l)[i]], x1 = pv$end[pv$chr == names(l)[i]],
         y0 = pv$cov[pv$chr == names(l)[i]], lwd = 3, col = "blue")
}

pv <- read.table("~/Projects/Botocudos/Files/Microbiome/Virome/Parvovirus/2019_03_14/geno3.txt")
colnames(pv)<- c("chr", "start", "end", "cov")
pv <- pv[pv$cov > 0,]

chrs <- unique(pv$chr)
l <- c(5017, 5028, 5028)
names(l) <- c("AY083234.1", "AJ249437.1", "NC_004295.1")

for(i in seq(1, length(l))){
  plot(pv$start[pv$chr == names(l)[i]], pv$cov[pv$chr == names(l)[i]], 
     type = "n" , xlim = c(0, l[i]), bty = "n",
     main = paste("Genotype 3, strain", names(l)[i]),
     xlab = "Position (bp)", ylab = "Genome coverage")
segments(x0 =pv$start[pv$chr == names(l)[i]], x1 = pv$end[pv$chr == names(l)[i]],
         y0 = pv$cov[pv$chr == names(l)[i]],lwd = 3, col = "brown")
}



```

# Bacteria
We remove variables (bacteria) for which we have more than 50% of missing data.

```{r}
# PCA on Phyla level


extract_counts <- function(rank){
  whole <- data.frame()
  
  for(ind in boto$MA){
    path <- paste("~/Projects/Botocudos/Files/Microbiome/Centrifuge/2019_03_07/centrifuge/", ind, "_nt.report", sep = "")
    report <- read.csv(path, sep = "\t", header = F)
    colnames(report) <- c("Reads_clade_percent",
                          "Reads_clade",
                          "Reads_taxon",
                          "Rank", 
                          "TaxID",
                          "Sci_name")
    report$Rank <- factor(report$Rank, 
                          levels = c("-", "S", "G", "F", "O", "C", "P", "K", "D"),
                          ordered = T)
    total_classified <- report$Reads_clade[report$TaxID == 1]
    
    # Index to prokkaryotes and viruses
    index <- grep("D", report$Rank)
    report <- report[c(index[1]:index[2]-1, index[3]:dim(report)[1]),]
    report$Reads_clade <- report$Reads_clade/total_classified
    colnames(report) <- sub("Reads_clade", ind, colnames(report))
    report$Sci_name <- sub("^ +", "", report$Sci_name)
    if(dim(whole)[1] == 0){
      whole <- report[report$Rank == rank, c("TaxID", "Sci_name", ind)]
    }else{
      whole <- join(whole, 
                    report[report$Rank == rank, c("TaxID", "Sci_name",ind)], 
                    by = "TaxID", type = "full")
    }
  }
  
  #hist(rowSums(is.na(whole)), breaks = dim(whole)[1])
  # remove rows with a lot of missing data
  whole <- whole[-which(rowSums(is.na(whole))/dim(whole)[2] > 0.5),] 
  whole[is.na(whole)] <- 0
  
  return(whole)
}

boto <- read.csv("~/Projects/Botocudos/Files/database_names.csv")
boto <- boto[-which(boto$Ethn == "Quack" |boto$MN == "MN0008"),]

```

```{r, fig.width=14}
ranks <- factor(c("-", "S", "G", "F", "O", "C", "P", "D", "K"), 
                  levels = c("-", "S", "G", "F", "O", "C", "P", "D", "K"),
                  ordered = T)

names(ranks) <- c("-", "Species", "Genus", "Family", "Order", "Phyla", "Domain")
for(rank in c("Phyla", "Order", "Family", "Genus", "Species")){
  cat(paste("## ", rank))
  
  for(do_log in c(F, T)){

    r <- ranks[rank]
    #whole <- extract_counts(r)
    whole <- read.csv(paste("~/Projects/Botocudos/Files/Microbiome/Centrifuge/2019_03_07/", rank, ".txt", sep = ""), header = T, sep = "\t")
    data <- t(as.matrix(whole[,3:dim(whole)[2]]))
    data[is.infinite(data)] <- 0
    rownames(data) <- ma2mn(rownames(data))
    data <- data[-c(which(rownames(data) == "MN00013")),]
    
    colnames(data) <- whole$Sci_name
    if(do_log){
      x <- prcomp(log(data+1e-10))
    }else{
      x <- prcomp(data)
    }
    
    
    p <- fviz_pca_ind(x, repel = T) 
    p <- p + 
      geom_point(x = x$x[rownames(x$x) == "MN01701",1], 
                 y = x$x[rownames(x$x) == "MN01701",2], col = "red") + 
      geom_point(x = x$x[rownames(x$x) == "MN1943",1], 
                 y = x$x[rownames(x$x) == "MN1943",2], col = "blue")
    
    y <- list(contrib = 10)
    contrib <- fviz_pca_var(x, select.var = y, repel = T, col.var = "contrib")
    
    var_exp <- fviz_screeplot(x)
    print(plot_grid(p, contrib, ncol = 2))
    print(var_exp)
  }
  
}
```

# PCA on pathogenic species

Finally, I decided to restrict the PC analysis to 
species found in humans. This has been a difficult task, and 
for now I downloaded information for species reported as pathogens
to humans from the PATRIC database.


```{r, eval =T, fig.width=15}
patric <- read.csv("~/Projects/Botocudos/Files/Microbiome/PATRIC/PATRIC_genome.csv")
colnames(patric) <- sub("NCBI.Taxon.ID", "TaxID", colnames(patric))

h <- list()
for(rank in c("Species")){
  cat(paste("## ", rank))
  for(do_log in c(F, T)){
    
    r <- ranks[rank]
    #whole <- extract_counts(r)
    whole <- read.csv(paste("~/Projects/Botocudos/Files/Microbiome/Centrifuge/2019_03_07/", rank, ".txt", sep = ""), header = T, sep = "\t")
    
    patho <- join(whole, patric[, c("TaxID", "Genome.ID")],
                  by = "TaxID")
    patho <- patho[complete.cases(patho),]
    patho$Genome.ID <- NULL
    patho <- unique(patho)
    data <- t(as.matrix(patho[,3:(dim(patho)[2])]))
    data[is.infinite(data)] <- 0
    rownames(data) <- ma2mn(rownames(data))
    data <- data[-c(which(rownames(data) == "MN00013")),]
    
    colnames(data) <- patho$Sci_name
    if(do_log){
      x <- prcomp(log(data+1e-10))
    }else{
      x <- prcomp(data)
    }
    
     heatmap(t(data), col = rev(heat.colors(93)))
    
    p <- fviz_pca_ind(x, repel = T) 
    p <- p + 
      geom_point(x = x$x[rownames(x$x) == "MN01701",1], 
                 y = x$x[rownames(x$x) == "MN01701",2], col = "red") + 
      geom_point(x = x$x[rownames(x$x) == "MN1943",1], 
                 y = x$x[rownames(x$x) == "MN1943",2], col = "blue")
    
    y <- list(contrib = 10)
    contrib <- fviz_pca_var(x, select.var = y, repel = T, col.var = "contrib")
    
    var_exp <- fviz_screeplot(x)
    print(plot_grid(p, contrib, ncol = 2))
    print(var_exp)
  }
  
}

```

```{r}

for(i in seq(3,25)){
  a <- (whole[,c(2,i)])
  a <- a[order(a[,2], decreasing = T),]
  #print(ma2mn(colnames(whole)[i]))
  #print(head(a, 10))
  #n <- as.character(ma2mn(as.character(colnames(whole)[i])))
  n <- colnames(whole)[i]
  name <- paste("~/Projects/Botocudos/Files/Microbiome/Centrifuge/", n, "_most_abundant.txt", sep = "")
  #write.table(a, file = name, sep = "\t", col.names = F, row.names = F, quote = F)
  print(a[which(a[,2]*15e6 > 100000),])
}
```