---
title: "Supplementary tables for mapping statistics and contamination"
author: "DI Cruz DÃ¡valos"
date: "2/5/2020"
output: 
  html_document:
      toc: true
      toc_float: true
      collapsed: false
      smooth_scroll: false
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, message = F, warning = F)
```


```{r}
require(plyr)
require(ggplot2)
require(cowplot)
require(tidyr)
require(dplyr)
require(scales)
require(knitr)
require(DT)
require(openxlsx)
```

```{r}


#------------------------------------------------------------------------------#
# Load isotopic data if available
backbone <- read.csv("~/Projects/Botocudos/Files/Tables/Backbone_Table1_2020.csv", stringsAsFactors = F)
colnames(backbone) <- c("Sample", "CalibratedDate", "Area", "State") 

```

```{r read in file}
# Whole or Trimmed
data_type <- "Whole"

myPath <- "~/Projects/Botocudos/Files/Summaries/2019_11_21/"
dp <- "sample_depth.hg19.csv"
st <- "sample_stats.hg19.csv"

dp <- read.csv(paste(myPath, data_type, "/", dp, sep = ""), stringsAsFactors = F)
dp$SM <- sub(".hg19", "", dp$Sample)
st <- read.csv(paste(myPath, data_type,"/", st, sep = ""), stringsAsFactors = F)

myStats <- join(st, dp, by = "SM")
myStats <- myStats[order(myStats$endo_final_prop),]
myStats$SM <- factor(myStats$SM, levels = unique(myStats$SM), ordered = T)

#------------------------------------------------------------------------------#
# Function to do take care of the mess for splitting the sample into  a mix of 
# libraries or separate libraries
do_bricolage <- function(sample, libs1, libs2, new_sm1, new_sm2, new_lb1, new_lb2,
                         path_depth_lib, lib_st){
  
  myDepth <- read.csv(path_depth_lib)
  
  myDepth$Sample <- sub(".depth.txt", "", myDepth$Sample)
  
  subset_stats <- function(lib_st, sample, libs, new_sample, lib_name){
    sample_stats <- lib_st[lib_st$SM == sample,]
  
    subset_index <-  sample_stats$LB %in% libs
    mySubset_stats <- data.frame(
                       SM = paste(sample, lib_name, sep = "_"),
                       SM_new = new_sample,
                       reads_raw = sum(sample_stats$reads_raw[subset_index]),
                       reads_trim = sum(sample_stats$reads_trim[subset_index]),
                       mapping = sum(sample_stats$mapping[subset_index]),
                       duplicates = sum(sample_stats$duplicates[subset_index]),
                       mapping_final = sum(sample_stats$mapping_final[subset_index])
    )
    mySubset_stats$trim_prop <- mySubset_stats$reads_trim/mySubset_stats$reads_raw*100
    mySubset_stats$endo_prop <- mySubset_stats$mapping/mySubset_stats$reads_raw*100
    mySubset_stats$duplicates_prop <- mySubset_stats$duplicates/mySubset_stats$mapping*100
    mySubset_stats$endo_final_prop <- mySubset_stats$mapping_final/mySubset_stats$reads_raw*100
    
    
    return(mySubset_stats)
  }

  set1 <- subset_stats(lib_st = lib_st, sample = sample, libs = libs1,
                       new_sample = new_sm1, lib_name = new_lb1)
  set2 <- subset_stats(lib_st = lib_st, sample = sample, libs = libs2,
                       new_sample = new_sm2, lib_name = new_lb2)
  
  sample_stats <- rbind(set1, set2)
  myDepth$SM <- myDepth$Sample
  
  mySample <- join(sample_stats, myDepth, by = "SM")
  mySample$SM <- mySample$SM_new
  mySample$Sample <- mySample$SM_new
  mySample$SM_new <- NULL
  mySample$LB <- NULL
  mySample$Library <- NULL
  return(mySample)
}

#------------------------------------------------------------------------------#
# Add MN0008_L3U and MN0008_non_U,
# MN1943_L1 and MN01701_L1L2 (remove mtCapture)

lib_st <- read.csv(paste("~/Projects/Botocudos/Files/Summaries/2019_11_21/",
                   data_type, "/library_stats.hg19.csv", sep = ""))
mn8 <- do_bricolage(sample = "MN0008", libs1 = c("L3U"), libs2 = c("L1", "L2"),
                    new_sm1 = "MN0008_L3U", new_sm2 = "MN0008_non_U", 
                    new_lb1 = "L3U", new_lb2 = "non_U",
                    path_depth_lib = paste("~/Projects/Botocudos/Files/Summaries/2019_11_21/",
                    data_type, "/MN0008_bylib.depth.txt", sep = ""),
                    lib_st = lib_st)

mn1943 <- do_bricolage(sample = "MN1943", libs1 = c("L1"), libs2 = c("mtCapture"),
                    new_sm1 = "MN1943", new_sm2 = "MN1943_mtCapture", 
                    new_lb1 = "L1", new_lb2 = "mtCapture",
                    path_depth_lib = paste("~/Projects/Botocudos/Files/Summaries/2019_11_21/",
                    data_type,"/MN1943_bylib.depth.txt", sep = ""),
                    lib_st = lib_st)

mn01701 <- do_bricolage(sample = "MN01701", libs1 = c("L1", "L2"), libs2 = c("mtCapture"),
                    new_sm1 = "MN01701", new_sm2 = "MN01701_mtCapture", 
                    new_lb1 = "L1L2", new_lb2 = "mtCapture",
                    path_depth_lib = paste("~/Projects/Botocudos/Files/Summaries/2019_11_21/",
                    data_type, "/MN01701_bylib.depth.txt", sep = ""),
                    lib_st = lib_st)


myStats <- myStats[!myStats$SM == "MN1943", ]
myStats <- myStats[!myStats$SM == "MN01701", ]
myStats <- rbind(myStats, mn8)
myStats <- rbind(myStats, mn1943[mn1943$SM == "MN1943",])
myStats <- rbind(myStats, mn01701[mn01701$SM == "MN01701",])

#------------------------------------------------------------------------------#
# go from 0 - 100 to 0 - 1 for proportions and further formatting
myStats$trim_prop <- myStats$trim_prop/100
myStats$duplicates_prop <- myStats$duplicates_prop/100
myStats$endo_prop <- myStats$endo_prop/100
myStats$endo_final_prop <- myStats$endo_final_prop/100

# format the 95CI of Ry
myStats$low_ry <-  sapply(strsplit(myStats$X95.CI, "-"), function(x) as.numeric(x[1]))
myStats$high_ry <-  sapply(strsplit(myStats$X95.CI, "-"), function(x) as.numeric(x[2]))

write.csv(myStats, 
          paste("~/Projects/Botocudos/Files/Summaries/2019_11_21/", 
                data_type, "/myStats.csv", sep = ""), 
          row.names = F)
```

## Supplementary Table 1:  Mapping stats per sample

$\sum^{numberreads}_{i=1}{l_i}/number_reads$

```{r define column names}

myColNames <- c("Sample",            "Sample",                          "National Museum of Brazil (Museu Nacional, MN) identifier",
                "SM",            "Sample",                              "National Museum of Brazil (Museu Nacional, MN) identifier",
                # "library",                        "Library", "DNA library identifier",
                # "LB",                        "Library", "DNA library identifier",
                
                "CalibratedDate",    "Calibrated date (AD)",            "OxCal estimated mean and standard deviation for calibrated dates",
                "Area",              "Region",                          "Sample origin (region) according to MN's archives",
                "State",             "State",                           "Sample origin (state) according to MN's archives", 
                
                "reads_raw",         "reads_raw",                       "Number of sequenced reads", 
                "reads_trim",        "reads_trim",                      "Number of reads retained after trimming",
                "trim_prop",         "trim_prop",                       "Percentage of reads retained after trimming",
                
                "mapping",          "mapping",                          "Number of reads mapped to the human genome (build 37.1) with a mapping quality >=30 before removing duplicates",
                "endo_prop",        "endo_prop",                        "Percentage of retained reads that were uniquely mapped to the human genome",
                
                "duplicates",        "duplicates",                      "Number of reads flagged as PCR duplicates by picardtools MarkDuplicates",
                "duplicates_prop",   "duplicates_prop",                 "Percentage of reads flagged as PCR duplicates by picardtools MarkDuplicates",
                 
                "mapping_final",     "mapping_final",                   "Number of reads mapped to the human genome (build 37.1) with a mapping quality >=30 after removing duplicates",
                "endo_final_prop",   "endo_final_prop",                 "Percentage of retained reads that were uniquely mapped to the human genome after removing duplicates",
                
                "AvgReadDepth_tot",  "DoC_endogenous",                  "Depth of coverage on human genome (number of unique bases divided by the length of the human genome reference build 37.1)",
                "AvgReadDepth_MT",   "DoC_MT",                          "Depth of coverage on mitochondrial genome (number of unique bases divided by the length of the mitochondrial genome, build 37.1)",
                
                "R_y",               "Ry_95CI",                         "Ratio of the reads mapping to the Y-chromosome and the reads mapping to sexual chromosomes (X and Y) and 95% confidence interval",
                "Sex",               "Sex",                             "Genomic sex as determined by the Ry ratio",
                "AvgReadLength",     "length_reads_endogenous",         "Average length of mapped reads with mapping quality >=30 after removing duplicates"
                )

change_name <- function(name){
  new_name <- myColNames[which(myColNames == name)[1] + 1]
  ifelse(is.na(new_name), return(name), return(new_name))
}

```

```{r prepare caption}
myCaption <- paste(unique(paste(myColNames[seq(2, length(myColNames), 3)], 
                   myColNames[seq(3, length(myColNames), 3)], sep = ": ")),
                   collapse = "\n\n")
```

```{r prepare numbers format}
# colname, accuracy, percent
myNumFormat <- matrix(c("reads_raw", 1, 0, 
                 "reads_trim", 1, 0,
                 "trim_prop", 0.1, 1,
                 "mapping", 1, 0,
                 
                 "endo_prop", 0.01, 1,
                 "duplicates", 1, 0,
                 "duplicates_prop", 0.1, 1,
                 
                 "mapping_final", 1, 0,
                 "endo_final_prop", 0.01, 1,
                 "DoC_endogenous", 0.001, 0,
                 "DoC_MT", 0.1, 0,
                 "length_reads_endogenous", 0.1, 0), ncol = 3, byrow = T)

format_column_num <- function(col_name, percent, accuracy, myT1){
  if(percent){
    myT1[,col_name]  <- percent(myT1[,col_name], accuracy = accuracy)
  }else{
    myT1[,col_name] <- number(myT1[,col_name], accuracy = accuracy)
  }
  return(myT1)
}

```

```{r prepare table}
myT1 <- myStats
myT1$R_y <-  paste(round(myT1$R_y, 3), " (", round(myT1$low_ry, 3), 
                   "-",  round(myT1$high_ry, 3),
                   ")", sep = "")
myT1$low_ry <- NULL
myT1$high_ry <- NULL

myT1 <- join(backbone, myT1, by = "Sample")

colnames(myT1) <- sapply(colnames(myT1), function(name) change_name(name))

samples_order <- myT1$Sample[order(myT1$DoC_endogenous)]

myT1$Sample <- factor(myT1$Sample, levels = samples_order, ordered = T)
myT1 <- myT1[order(myT1$Sample, decreasing = T),]

# Drop duplicated column
myT1[,which(colnames(myT1) == "Sample")[2]] <- NULL 
# Drop some columns
myT1$SE <- NULL
myT1$Nseqs <- NULL
myT1$NchrY <- NULL
myT1$NchrY.NchrX <- NULL
myT1$X95.CI <- NULL

myXL <- myT1
```

```{r format prettyNum}
for(i in seq(1, nrow(myNumFormat))){
  myT1 <- format_column_num(myNumFormat[i, 1], 
                            as.numeric(myNumFormat[i,3]), 
                            as.numeric(myNumFormat[i,2]),
                            myT1)
}
```

```{r}
# kable(myT1, caption = myCaption, digits = 1, row.names = F)

myT1 %>%
DT::datatable(
    extensions = 'Buttons', 
    options = list(dom = 'Bfrtip', 
                   buttons = list('excel', "csv")),
    rownames = F, caption = myCaption)
```

```{r write myStats}
last_row_NA <- sapply(myT1[1,], function(x) return(NA))
myExplanation <- unique(paste(myColNames[seq(2, length(myColNames), 3)], 
                    myColNames[seq(3, length(myColNames), 3)], sep = ": "))

myT1 <- rbind(myT1, last_row_NA)
for(new_row in myExplanation){
  last_row_NA <- as.data.frame(t(sapply(myT1[1,], function(x) return(NA))))
  last_row_NA$Sample <- new_row
  myT1 <- rbind(myT1, last_row_NA)
}

write.table(myT1, paste("~/Projects/Botocudos/Files/Tables/Sup_T1_", data_type,
                        ".csv", sep = ""), 
            na = "",
            row.names = F, col.names = T, sep = ",")
```

```{r save to excel with openxlsx}
## data.frame to write
df <- myXL


## Formatting can be applied simply through the write functions 
#==============================================================================#
## create a workbook and add a worksheet
wb <- createWorkbook()
addWorksheet(wb, paste("Sup_T1_", data_type, sep = ""))
#--------------
## headerStyles
hs1 <- createStyle(fgFill = "white", halign = "CENTER", textDecoration = "Bold",
                   border = c("top", "bottom", "left", "right"),
                   borderStyle = "thin",
                   fontColour = "black", fontSize = 16)

showGridLines(wb, 1, showGridLines = FALSE)
setColWidths(wb, 1, cols = 1:ncol(myT1), widths = "auto")
#-------------
# write table
writeData(wb, 1, df, startRow = 1, startCol = 1, headerStyle = hs1,
          borders = "columns", borderStyle = "thin")
#-------------------
# add column styles
# percent
percent_cols <- which(colnames(df) %in% c("trim_prop", "duplicates_prop"))
s <- createStyle(numFmt = "0.0%")
addStyle(wb, 1, style = s,  cols = percent_cols, 
         rows = 1:nrow(df)+1, gridExpand = TRUE, stack = T)
# thousands separator
comma_cols <- which(colnames(df) %in% c("reads_raw", "reads_trim", "mapping", 
                                        "duplicates", "mapping_final"))
s <- createStyle(numFmt = "COMMA")
addStyle(wb, 1, style = s, rows = 1:nrow(df)+1, cols = comma_cols,
         gridExpand = TRUE, stack = T)
# trim decimal places and format %
decimal_cols <- which(colnames(df) %in% c("endo_prop", "endo_final_prop"))
s <- createStyle(numFmt = "#0.00%")
addStyle(wb, 1, style = s, rows = 1:nrow(df)+1, cols = decimal_cols,
         gridExpand = TRUE, stack = T)
# trim decimal places
decimal_cols <- which(colnames(df) %in% c("DoC_MT", "length_reads_endogenous"))
s <- createStyle(numFmt = "0.0")
addStyle(wb, 1, style = s, rows = 1:nrow(df)+1, cols = decimal_cols,
         gridExpand = TRUE, stack = T)

#format data font type
theCaption <- myT1[is.na(myT1$reads_raw),]
s <- createStyle(fontSize = 14)
addStyle(wb, 1, style = s, rows = 1:(nrow(df)+nrow(theCaption))+1, cols = 1:ncol(df),
         gridExpand = TRUE, stack = T)
#------------------------------------------------------------------------------#
# write column explanation
writeData(wb, 1, theCaption, startRow = nrow(df)+2,
          startCol = 1, colNames = F, 
          borders = "none")

## writing as an Excel Table
openXL(wb) ## opens a temp version
saveWorkbook(wb, paste("~/Projects/Botocudos/Files/Tables/Sup_T1_",
                       data_type, ".xlsx", sep = ""), 
             TRUE)
```


```{r make a plot}
myStats <- myStats[order(myStats$AvgReadDepth_tot),]
stats_to_plot <- myStats %>% 
  mutate(log_reads_raw = log10(reads_raw),
         log_reads_trim = log10(reads_trim),
         log_mapping = log10(mapping),
         log_duplicates = log10(duplicates),
         log_mapping_final = log10(mapping_final),
         log_endo_prop = log10(endo_prop),
         log_endo_final_prop = log10(endo_final_prop)) %>% 
  gather( name, value,  -SM, -low_ry, -high_ry, -Nseqs, 
          -NchrY.NchrX, -NchrY, -SE, -Sex, -X95.CI, -Sample) 

# stats_to_plot$value <- as.numeric(stats_to_plot$value)

stats_to_plot$name <- factor(stats_to_plot$name, levels = c("reads_raw",
                                                            "log_reads_raw",
                                                            "reads_trim",
                                                            "log_reads_trim",
                                                            "trim_prop",
                                                            "mapping",
                                                            "log_mapping",
                                                            "duplicates",
                                                            "log_duplicates",
                                                            "duplicates_prop",
                                                            "mapping_final",
                                                            "log_mapping_final",
                                                            "R_y",
                                                            "endo_prop",
                                                            "log_endo_prop",
                                                            "endo_final_prop",
                                                            "log_endo_final_prop",
                                                            "AvgReadDepth_tot",
                                                            "log_AvgReadDepth_tot",
                                                            "AvgReadDepth_MT",
                                                            "log_AvgReadDepth_MT",
                                                            "AvgReadLength"),
                             ordered = T)

####
# y-axis to meaningful numbers (i.e., 10^10, 10^4, etc...)

# one plot in linear scale, another in log

stats_to_plot %>%
ggplot(  aes(x = SM, y = value)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  theme_minimal()+
  theme(axis.text.x = element_text(angle = 90, size = 10),           
        
        strip.text.x = element_text(size = 12)) +
  facet_wrap(facets = vars(name), scales = "free_y",
             ncol = 3)  +
  labs(x = NULL, y = NULL)
```

### Plots with summary statistics
```{r}

prepare_stats <- function(myPath){
  myStats <- read.csv(myPath, stringsAsFactors = F)
  myStats <- myStats[order(myStats$AvgReadDepth_tot),]
  myStats$SM <- factor(myStats$SM, levels = unique(myStats$SM), ordered = T)
  stats_to_plot <- myStats %>% 
    mutate(log_reads_raw = log10(reads_raw),
           log_reads_trim = log10(reads_trim),
           log_mapping = log10(mapping),
           log_duplicates = log10(duplicates),
           log_mapping_final = log10(mapping_final),
           log_endo_prop = log10(endo_prop),
           log_endo_final_prop = log10(endo_final_prop)) %>% 
    gather( name, value,  -SM, -low_ry, -high_ry, -Nseqs, 
            -NchrY.NchrX, -NchrY, -SE, -Sex, -X95.CI, -Sample) 
  
  # stats_to_plot$value <- as.numeric(stats_to_plot$value)
  
  stats_to_plot$name <- factor(stats_to_plot$name, 
                               levels = c("reads_raw", "log_reads_raw",
                                          "reads_trim", "log_reads_trim", "trim_prop",
                                          "mapping", "log_mapping",
                                          "duplicates", "log_duplicates", "duplicates_prop",
                                          "mapping_final", "log_mapping_final",
                                          "R_y",
                                          "endo_prop", "log_endo_prop",
                                          "endo_final_prop", "log_endo_final_prop",
                                          "AvgReadDepth_tot", "log_AvgReadDepth_tot",
                                          "AvgReadDepth_MT", "log_AvgReadDepth_MT",
                                          "AvgReadLength"),
                               ordered = T)
  return(stats_to_plot)
}
####
# y-axis to meaningful numbers (i.e., 10^10, 10^4, etc...)

# one plot in linear scale, another in log
path_trim <- "~/Projects/Botocudos/Files/Summaries/2019_11_21/Trimmed/myStats.csv"
path_whole <- "~/Projects/Botocudos/Files/Summaries/2019_11_21/Whole/myStats.csv"

whole <- prepare_stats(path_whole)
whole$type <- "Untrimmed"
trim <- prepare_stats(path_trim)
trim$type <- "Trimmed"
myDF <- rbind(whole, trim)

myDF$type <- factor(myDF$type, levels = c("Untrimmed", "Trimmed"), ordered = T)

myDF %>%
ggplot(  aes(x = SM, y = value, fill = type)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  theme_minimal()+
  theme(axis.text.x = element_text(angle = 90, size = 10),           
        strip.text.x = element_text(size = 12)) +
  facet_wrap(facets = vars(name), scales = "free_y",
             ncol = 3)  +
  labs(x = NULL, y = NULL) +
  scale_fill_discrete()



for(myCol in unique(myDF$name)){
  myPath <- paste("~/Projects/Botocudos/Plots/Sequencing/2020_02_25/comparison_",
                  myCol, ".png", sep = "")
  
  p <- ggplot(myDF[myDF$name == myCol, ],aes(x = SM, y = value, fill = type)) +
    geom_bar(stat = "identity", position = position_dodge()) +
    labs(title = myCol) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 90, size = 16, hjust = 1),           
          axis.text.y = element_text(size = 14),
          plot.title = element_text(size = 18, face = "bold", hjust = 0.5)) +
    labs(x = NULL, y = NULL) +
    scale_fill_discrete() 
  png(myPath, width = 9, height = 6, res = 250, bg = NA, units = "in")
  print(p)
  dev.off()
}

for(myCol in unique(whole$name)){
  myPath <- paste("~/Projects/Botocudos/Plots/Sequencing/2020_02_25/whole_",
                  myCol, ".png", sep = "")
  
  p <- ggplot(whole[whole$name == myCol, ],aes(x = SM, y = value, fill = type)) +
    geom_bar(stat = "identity", position = position_dodge(), fill = "salmon") +
    labs(title = myCol) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 90, size = 16, hjust = 1),           
          axis.text.y = element_text(size = 14),
          plot.title = element_text(size = 18, face = "bold", hjust = 0.5)) +
    labs(x = NULL, y = NULL) 
    
  png(myPath, width = 9, height = 6, res = 250, bg = NA, units = "in")
  print(p)
  dev.off()
}


for(myCol in unique(trim$name)){
  myPath <- paste("~/Projects/Botocudos/Plots/Sequencing/2020_02_25/trim_",
                  myCol, ".png", sep = "")
  
  p <- ggplot(trim[trim$name == myCol, ],aes(x = SM, y = value, fill = type)) +
    geom_bar(stat = "identity", position = position_dodge(), fill = "turquoise3") +
    labs(title = myCol) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 90, size = 16, hjust = 1),           
          axis.text.y = element_text(size = 14),
          plot.title = element_text(size = 18, face = "bold", hjust = 0.5)) +
    labs(x = NULL, y = NULL) 
    
  png(myPath, width = 9, height = 6, res = 250, bg = NA, units = "in")
  print(p)
  dev.off()
}
```

```{r venn diagram}
user <- rep("non_USER", 24)
lib_strategy <- rep("single-end", 24)
enrichment <- rep("not enriched", 24 )
tissue <- rep("teeth", 24)
ethnicity <- rep("Botocudo", 24)
carbon <- rep("post-Contact", 24)

# 1: MN0008
# 2: MN1943
# 3: MN01701
# 4: MN00019
user[1] <- "USER"
lib_strategy[2:3] <- "paired-end"
enrichment[2:3] <- "enriched"
tissue[1] <- "petrous"
tissue[4] <- "skull"
ethnicity[3] <- "Mummy"
ethnicity[4] <- "Sambaqui"
carbon[2:3] <- "pre-contact"

```

```{r}
non_user <- myStats$Sample[-which(myStats$Sample %in% c("MN0008_L3U", "MN0008_non_U"))]
single <- non_user
tissue <- non_user

paired <- c("MN1943", "MN01701")

enriched <- paired
post_contact <- non_user[-which(non_user %in% paired)]
pre_contact <- paired
ethnicity <- post_contact

require(VennDiagram)

# Prepare a palette of 3 colors with R colorbrewer:
require(RColorBrewer)
myCol <- brewer.pal(8, "Pastel2")

myList <- list(non_user, single, paired, 
               enriched, post_contact, 
               pre_contact, ethnicity)
myNames <- c("Non-USER", "Single-End", "Paired-End", 
             "Enriched", "Post-Contact", 
             "Pre-Contact", "Botocudo")
venn.diagram(
        x = myList,
        category.names = myNames,
        filename = '~/Desktop/venn_diagramm.png',
        output=TRUE),
        
        # Output features
        imagetype="png" ,
        height = 480 , 
        width = 480 , 
        resolution = 300,
        compression = "lzw",
        
        # Circles
        lwd = 2,
        lty = 'blank',
        fill = myCol,
        
        # Numbers
        cex = .6,
        fontface = "bold",
        fontfamily = "sans",
        
        # Set names
        cat.cex = 0.6,
        cat.fontface = "bold",
        cat.default.pos = "outer",
        # cat.pos = c(-27, 27, 135),
        # cat.dist = c(0.055, 0.055, 0.085),
        cat.fontfamily = "sans",
        rotation = 1
)
```

## Supplementary Table 2: Mapping stats per library

```{r prepare number format}
myNumFormat <- matrix(c("reads_raw", 1, 0, 
                 "reads_trim", 1, 0,
                 "duplicates", 1, 0,
                 "duplicates_prop", 3, 1,
                 "mapping_final", 1, 0,
                 "endo_prop", 4, 1
                 ), ncol = 3, byrow = T)
```

```{r}
lib_st <- read.csv("~/Projects/Botocudos/Files/Summaries/2019_11_21/Whole/library_stats.hg19.csv")
whole_st$LB <- "All"

myT2 <- rbind(whole_st, lib_st)
myT2 <- myT2[order(myT2$SM, myT2$LB),]
myT2$LB <- factor(myT2$LB, levels = c("mtCapture", "L3U", "L2", "L1", "All"),
                  ordered = T)
myT2$SM <- factor(myT2$SM, levels = samples_order, ordered = T)
myT2 <- myT2[order(myT2$SM, myT2$LB, decreasing = T),]



myT2 <- myT2[,c("SM", "LB","reads_raw", "reads_trim", "trim_prop", 
                "duplicates", "duplicates_prop", "mapping","mapping_final",
                "endo_prop")]

myT2_final <- myT2
myT2_final <- myT2_final[, c("SM", "LB", "reads_raw", "reads_trim", "trim_prop",
                             "duplicates", "duplicates_prop", "mapping_final", "endo_prop")]
# colnames(myT2_final) <- sapply(colnames(myT2_final), function(name) change_name(name))
# myT2_final$percent_duplicated_reads <- myT2_final$percent_duplicated_reads/100
# myT2_final$percent_unique_reads_endogenous <- myT2_final$percent_unique_reads_endogenous/100

for(i in seq(1, nrow(myNumFormat))){
  myT2_final <- format_column_num(myNumFormat[i, 1], 
                            as.numeric(myNumFormat[i,3]), 
                            as.numeric(myNumFormat[i,2]),
                            myT2_final)
}
```


```{r}
# kable(myT2_final, row.names = F, digits = 1)
myT2_final %>%
  DT::datatable(
    extensions = 'Buttons', 
    options = list(dom = 'Bfrtip', 
                   buttons = list('excel', "csv")),
    rownames = F)

```


## Supplementary Table 3: Contamination estimates


```{r add results from contaminationX}
#------------------------------------------------------------------------------#
## X
minDepth <- seq(2,3)
path <- "~/Projects/Botocudos/Files/Contamination/2020_01_24/contaminationX/"
mySamples <- unique(myT1$Sample)
libs <- c("All")

s <- mySamples[1]
l <- libs[1]
md <- minDepth[2]

empty_df <- function(sample, lib){
  myDF <- data.frame(method = c("One-cns", "Two-cns"),
                     estimate = c(NA, NA), lb = c(NA, NA), ub = c(NA, NA),
                     err = c(NA, NA), nSites = c(NA, NA))
  return(myDF)
}

x_cont <- data.frame()
for(s in mySamples){
  for(l in libs){
    for(md in minDepth){
      
      tmp_res <- tryCatch(tmp_res <- read.table(paste(path, 
                                                      s,".hg19_", l, 
                                                      "_md", md, ".result", sep = "")),
                          error =  function(e) empty_df(s, l),
                          finally = function() return(tmp_res))
      
      colnames(tmp_res) <- c("method", "estimate", "lb", "ub", "err", "nSites")
      tmp_res$Sample <- s
      tmp_res$library <- l
      tmp_res$depth <- md
      x_cont <- rbind(x_cont, tmp_res)
    }
  }
}

x_cont <- join(x_cont, myT1[,c("Sex", "Sample")], by = "Sample")
x_cont <- x_cont[!is.na(x_cont$estimate) & x_cont$Sex %in% c("XY", "Not Assigned", "consistent with XY but not XX"),]

males <- x_cont[x_cont$method == "Two-cns",]

# write.table(males, "~/Projects/Botocudos/Files/Contamination/2020_01_24/contaminationX/estimates_X_males.txt", 
# sep = "\t", quote = F, col.names = T, row.names = F)

males_fmt <- males

males_fmt$estimate <- percent(males_fmt$estimate)
males_fmt$lb <- percent(males_fmt$lb)
males_fmt$ub <- percent(males_fmt$ub)

colnames(males_fmt) <- c("X_method", "X_estimate", "X_low_est", "X_upper_est",
                         "X_error", "X_num_sites", "sample", "library", "X_min_depth",
                         "sex")


```

```{r, eval = F}
kable(males_fmt, row.names = F)
```

```{r add results from contammix}
prefix <- "~/Projects/Botocudos/Files/Contamination/2020_01_24/contamMix/"
samples <- samples_order
rmTrans <- c("all", "rmTrans")

# We ran the chains for 100k iterations, 
# we will subsample a fraction of these samples to plot the posterior distribution

subsample_contammmix <- function(path, sample, rmTrans, size = 100, lib = "All",
                                 burnIn = 0.1){

    myRda <- paste(path, sample, "/", s, ".hg19_", lib, "_", rmTrans, ".Rdata", sep = "")

  print(myRda)
  load(myRda)
  notBurnIn <- as.integer(burnIn*size)
  tiny <- # sample(
    #remove burnin
    c(res$chains[[1]][size - ((size- notBurnIn):0),2])#,
                 #res$chains[[2]][,2], res$chains[[3]][,2])#, size)
  result <- data.frame(estimate = tiny, damage = rmTrans, lib = lib,
                       sample = sample)
  return(result)
}

# contammix <- data.frame()
# for(s in samples){
#   libs <- (myT2$LB[myT2$SM == s])
#   for(l in libs){
#     for(r in rmTrans){
#       tmp_contam <- subsample_contammmix(prefix, s, r, 100000, l)
#       contammix <- rbind(contammix, tmp_contam)
#     }
#   }
# }


#------------------------------------------------------------------------------#
# Order libraries and save (or load) results
# contammix$lib <- factor(contammix$lib,
#                         levels = c("L1", "L2", "L3U", "mtCapture","All"), 
#                         ordered = T)
# contammix$coverage <- 0
# for(s in unique(contammix$sample)){
#   for(l in unique(contammix$lib[contammix$sample == s])){
#     contammix$coverage <- boto$hits_coverage_mitochondrial[boto$sample == s & boto$library == l]
#   }
# }

# save(contammix, file="~/Projects/Botocudos/Files/Contamination/2020_01_24/contamMix//contamination.Rdata")
load("~/Projects/Botocudos/Files/Contamination/2020_01_24/contamMix/contamination.Rdata")


path <- "~/Projects/Botocudos/Files/Contamination/2020_01_24/contamMix/"
rmTrans <- c("all", "rmTrans")
annot_mito <- data.frame()

for(s in samples){
  libs <- (myT2$LB[myT2$SM == s])
  for(lib in libs){
    for(r in rmTrans){
      myRda <- paste(path, s, "/", s, ".hg19_", lib, "_", r, ".Rdata", sep = "")
      #print(myRda)
      load(myRda)
      nReads <- dim(res$mnMatrix)[1]
      # coverage <- boto$hits_coverage_mitochondrial[boto$sample == s & boto$library == lib]
      annot_tmp <- data.frame(nReads = nReads, #coverage = coverage, 
                              damage = r, lib = lib,
                           sample = s)
      
      annot_mito <- rbind(annot_mito, annot_tmp)
    }
  }
}



MAP <- read.csv("~/Projects/Botocudos/Files/Contamination/2020_01_24/contamMix/estimates.txt", header = F)

colnames(MAP) <- c("sample", "lib", "damage","map", "low", "high")
MAP$lib <- sub("S1", "L1", MAP$lib)
annot_mito <- join(annot_mito, MAP, by = c("sample", "lib", "damage"))

```

```{r print estimates from contamMix, eval = F}

kable(mito, row.names = F)
```

```{r add results from schmutzi}
path <- "~/Projects/Botocudos/Files/Contamination/2020_01_24/schmutzi/"

schmutzi_files <- list.files(path, pattern =".est")

schmutzi <- data.frame()

for(f in schmutzi_files){
  tmp_schmutzi <- read.table(paste(path, f, sep =""))
  colnames(tmp_schmutzi) <- c("estimate", "low", "up")
  tmp_schmutzi$sample <- strsplit(f, "_")[[1]][1]
  schmutzi <- rbind(schmutzi, tmp_schmutzi)
  
}
```

```{r print estimates from schmutzi, eval = F}
kable(schmutzi, row.names = F)
```

```{r}
myAccuracy <- 1

# contamMix
mito <- annot_mito[annot_mito$lib == "All" & annot_mito$damage == "all",]
mito$damage <- NULL
mito$lib <- NULL
colnames(mito) <- c("contamMix_num_reads", "Sample", 
                    "contamMix_estimate", "contamMix_low", "contamMix_up")

mito$contamMix_estimate <- percent(1 - mito$contamMix_estimate, accuracy = myAccuracy)
mito$contamMix_low <- percent(1 - mito$contamMix_low, accuracy = myAccuracy)
mito$contamMix_up <- percent(1 - mito$contamMix_up, accuracy = myAccuracy)
mito$contamMix_estimate <- paste(mito$contamMix_estimate, 
                                 " (", mito$contamMix_up, 
                                 ", ", mito$contamMix_low, ")", sep = "")
mito <- mito[, c("Sample", "contamMix_num_reads", "contamMix_estimate")]


# schmutzi
mito_sch <- schmutzi
colnames(mito_sch ) <- c("schmutzi_estimate", "schmutzi_low", "schmutzi_up", "Sample")
mito_sch$schmutzi_estimate <- percent(mito_sch$schmutzi_estimate, accuracy = myAccuracy)
mito_sch$schmutzi_low <- percent(mito_sch$schmutzi_low, accuracy = myAccuracy)
mito_sch$schmutzi_up <- percent(mito_sch$schmutzi_up, accuracy = myAccuracy)
mito_sch$schmutzi_estimate <- paste(mito_sch$schmutzi_estimate, 
                                    " (", mito_sch$schmutzi_low, 
                                    ", ",mito_sch$schmutzi_up, ")", sep = "")
mito_sch <- mito_sch[,c("Sample", "schmutzi_estimate")]

# contaminationX
x <- males_fmt[males_fmt$X_min_depth == 3,]
x$library <- NULL
x$X_method <- NULL
x$sex <- NULL
x$X_error <- NULL
x$X_min_depth <- NULL
colnames(x) <- c("contaminationX_estimate", "contaminationX_low", 
                 "contaminationX_up", "contaminationX_num_sites", "Sample")
x$contaminationX_estimate <-  paste(x$contaminationX_estimate, 
                                    " (", x$contaminationX_low, 
                                    ", ",x$contaminationX_up, ")", sep = "")
x <- x[,c("Sample", "contaminationX_estimate")]

# Merge
cont <- join(myT1[,c("Sample", "Sex")], mito, by = "Sample")
cont <- join(cont, mito_sch, by = "Sample")
cont <- join(cont, x, by = "Sample")

# kable(cont, row.names = F)

cont %>%
  DT::datatable(
    extensions = 'Buttons', 
    options = list(dom = 'Bfrtip', 
                   buttons = list('excel', "csv")),
    rownames = F)
```

## Supplementary Table 4: Haplogroups?

!!!
Haplogroups called (whenever it was possible)
cleaning and gathering output from haplogrep and David's script
!!!

```{r parse mt-haplogroups}
myPath <- "~/Projects/Botocudos/Files/Haplogroups/2020_01_23/"
hap_files <- list.files(myPath, pattern = ".haplo$")
mt_hap <- data.frame()

for(f in hap_files){
  tmp_hap <- read.table(paste(myPath, f, sep = ""), header = T)
  s <- sub(".haplo", "", f)
  tmp_hap$Sample <- s
  mt_hap <- rbind(mt_hap, tmp_hap)
  }

mt_hap <- mt_hap[,c("Sample", "Haplogroup")]


y_hap <- data.frame(Sample = c("MN0003",
                               "MN00064",
                               "MN00067",
                               "MN0008",
                               "MN0009",
                               "MN00346"),
                    Haplogroup = c("Q1a2a1-L54",
                                   "Q1a2a1a1-M3",
                                   "Q1a2a1a1-M3",
                                   "Q1a2a1a1-M3",
                                   "Q1a2a1a1-M3",
                                   "Q1a2a1a1-M3"))

myHaplos <- join(myT1[,c("Sample", "Sex")], mt_hap, by = "Sample")
myHaplos <- join(myHaplos, y_hap, by = "Sample")

kable(myHaplos, row.names = F)
```

## Length

Lots of plots per library, merged in one pdf

```{r}
# Function to fill in gaps in data frame (counts equaal to zero)
fill_gaps <- function(le){
  mySeq <- seq(min(le$length), max(le$length))
  mySeq <- mySeq[!mySeq %in% le$length]
  myGap <- data.frame(length = mySeq, counts = 0)
  new_le <- join(le, myGap, by = c("length", "counts"), "full")
  new_le <- new_le[order(new_le$length),]
  return(new_le)
}
```

```{r load files per sample}
myPath <- paste("~/Projects/Botocudos/Files/bamdamage/2020_01_23/",
                data_type, sep = "")

samples <- as.character(myStats$SM[order(myStats$AvgReadDepth_tot, decreasing = T)])
if(!"MN0008_L3U" %in% samples){
  samples <- c(samples, "MN0008_L3U", "MN0008_non_U")  
}
myLength <- vector("list", length = length(samples))
names(myLength) <- samples

for(s in samples){
  path_length <- paste(myPath, s, sep = "/")
  path_length <- paste(path_length, "/", s, ".hg19.length.csv", sep = "")
  tmp_length <- read.csv(path_length)
  myLength[[s]] <- tmp_length
  }   


```

```{r plots per sample, eval = T, fig.width=16, fig.height=16}
myColor <- "royalblue"
nCol <- 6
samples_per_row <- ceiling(length(samples)/nCol)
cex <- 1.1

blank_plot <- function(){
  plot(1,1, type = "n", bty = "n", axes = F, xlab = NA, ylab = NA)
}

pdf(paste("~/Projects/Botocudos/Plots/Sequencing/Length_dist_",
          data_type,".pdf", sep = ""),
    width = 12, height = 12)
layout(matrix(seq(1, nCol*samples_per_row), ncol = nCol, byrow = T))

for(s in samples){
    le <- myLength[[s]]
    plot.title <- s
    myMean <- round(sum(le$length*le$counts)/sum(le$counts), 1)
    barplot(le$counts/sum(le$counts), space = 0,
         col = myColor, bty = "n", 
         xlab = "Length", ylab = "Frequency", border = NA,
         main = plot.title, lwd = 0.2)
    axis(1, at= seq(0.5, nrow(le), 5), 
         labels = seq(min(le$length), max(le$length), 5))
    text(20, 0.9*max(le$counts/sum(le$counts)), 
         labels = paste("Mean: ", myMean, " bp", sep = ""),
        cex = 0.8)
    abline(v = myMean-20, lty = "dashed")
}

dev.off()
```

```{r load files per library}
myPath <- paste("~/Projects/Botocudos/Files/bamdamage/2020_01_23/",
                data_type,"/", sep = "")

myLength <- vector("list", length = length(samples))
names(myLength) <- samples

for(s in samples){
  myLibs <- lib_st$LB[lib_st$SM == s]
  some_length <- vector("list", length = length(myLibs))
  names(some_length) <- myLibs
  
  for(l in myLibs){
    path_length <- paste(myPath, s, l, "library_bamdamage", sep = "/")
    path_length <- paste(path_length, "/", l, ".hg19.length.csv", sep = "")
    tmp_length <- read.csv(path_length)
    some_length[[l]] <- tmp_length
  }
  myLength[[s]] <- some_length
  }


```

```{r plots per library, eval = F}
myColor <- "royalblue"
nCol <- max(sapply(myLength, length))
samples_per_page <- 5
cex <- 1.2

blank_plot <- function(){
  plot(1,1, type = "n", bty = "n", axes = F, xlab = NA, ylab = NA)
}

pdf(paste("~/Projects/Botocudos/Plots/Sequencing/Length_dist_libs_",data_type,".pdf",sep = ""),
    width = 9, height = 12)
layout(matrix(seq(samples_per_page*nCol), nrow = samples_per_page, byrow = T))

for(s in samples){
  myLibs <- lib_st$LB[lib_st$SM == s]
  myRow <- which(samples == s)
  
  if(!myRow %% samples_per_page){
    # plot.new()
    layout(matrix(seq(samples_per_page*nCol), nrow = samples_per_page, byrow = T))  
  }
  
  for(l in myLibs){
    le <- myLength[[s]][[l]]
    # le <- le[le$counts >0,]
    
    plot.title <- paste(s, l)
    barplot(le$counts/sum(le$counts), space = 0,
         col = myColor, bty = "n", 
         xlab = "Length", ylab = "Frequency", #border = "white",
         main = plot.title, lwd = 0.2)
    axis(1, at= seq(0.5, nrow(le)-0.5, 5), 
         labels = seq(min(le$length), max(le$length), 5))
  }
  plots_left <- nCol - length(myLibs)
  while(plots_left > 0){
    blank_plot()
    plots_left <- plots_left - 1
  }
}

dev.off()



```

## Damage

Lots of plots per library, merged in one pdf


```{r load damage per sample}
myPath <- paste("~/Projects/Botocudos/Files/bamdamage/2020_01_23/", data_type, sep = "")

if(!"MN0008_L3U" %in% samples){
  samples <- c(samples, "MN0008_L3U", "MN0008_non_U")  
}

myDam <- vector("list", length = length(samples))
names(myDam) <- samples

for(s in samples){
  path_dam <- paste(myPath, s, sep = "/")
  path_dam <- paste(path_dam, "/", s, ".hg19.dam_5prime.csv", sep = "")
  tmp_dam5 <- read.csv(path_dam)
  path_dam <- sub("5prime", "3prime", path_dam)
  tmp_dam3 <- read.csv(path_dam)
  
  some_dam <- list(five_prime = tmp_dam5, three_prime = tmp_dam3)
  myDam[[s]] <- some_dam
}


```

```{r plot damage per sample, eval=T, fig.width=16, fig.height=16}
nCol <- 8
nRow <- 7
cex <- 1.2

blank_plot <- function(){
  plot(1:10,1:10, type = "n", bty = "n", axes = F, xlab = NA, ylab = NA)
}

pdf(paste("~/Projects/Botocudos/Plots/Sequencing/Damage_", data_type, ".pdf", sep = ""),
    width = 14, height = 12)
layout(matrix(seq(nCol*nRow), ncol = nCol, byrow = T))

for(s in samples){
    y2 <- 0.35
    plot.title <- s
    five_prime <- myDam[[s]][["five_prime"]]
    three_prime <- myDam[[s]][["three_prime"]]
    plot_bamdamage(five_prime = five_prime, three_prime = three_prime, 
                  cex = 1, 
                  y2 = y2, plot.title = plot.title,
                  myColor = c())
}
plot_bamdamage(legend_only = T,
                  myColor = c(),
               legend_cex = 1, ncol_legend = 3)

dev.off()
```


```{r load damage per library}
myPath <- paste("~/Projects/Botocudos/Files/bamdamage/2020_01_23/", data_type,"/", sep = "")

myDam <- vector("list", length = length(samples))
names(myDam) <- samples

for(s in samples){
  myLibs <- lib_st$LB[lib_st$SM == s]
  some_dam <- vector("list", length = length(myLibs))
  names(some_dam) <- myLibs
  
  for(l in myLibs){
    path_dam <- paste(myPath, s, l, "library_bamdamage", sep = "/")
    path_dam <- paste(path_dam, "/", l, ".hg19.dam_5prime.csv", sep = "")
    tmp_dam5 <- read.csv(path_dam)
    path_dam <- sub("5prime", "3prime", path_dam)
    tmp_dam3 <- read.csv(path_dam)
    
    some_dam[[l]] <- list(five_prime = tmp_dam5, three_prime = tmp_dam3)
  }
  myDam[[s]] <- some_dam
  }


```

```{r plot damage per library, eval=F}
nCol <- max(sapply(myDam, length))
samples_per_page <- 5
cex <- 1.2

blank_plot <- function(){
  plot(1:10,1:10, type = "n", bty = "n", axes = F, xlab = NA, ylab = NA)
}

pdf(paste("~/Projects/Botocudos/Plots/Sequencing/Damage_lib_", data_type, ".pdf", sep = ""),
    width = 12, height = 12)
layout(matrix(seq(samples_per_page*nCol*2), nrow = samples_per_page, byrow = T))

for(s in samples){
  myLibs <- lib_st$LB[lib_st$SM == s]
  myRow <- which(samples == s)
  
  if(!myRow %% samples_per_page){
    # plot.new()
    layout(matrix(seq(samples_per_page*nCol*2), nrow = samples_per_page, byrow = T))  
  }
  
  for(l in myLibs){
    y2 <- 0.35
    le <- myDam[[s]][[l]][["five_prime"]]
    
    plot.title <- paste(s, l)
    par(mar = c(4, 3, 3, 1))
    plot(le$C..T[1:25], type = "b",
         col = myColor_CT, bty = "n", 
         xlab = "5' position", ylab = "Frequency", lwd = 1.5, 
         main = plot.title, cex = cex, axes = F, ylim = c(0, y2))
    axis(side = 1, at = seq(0,25,5))
    axis(side=2)
    
    le <- myDam[[s]][[l]][["three_prime"]]
    par(mar = c(4,1,3,3))
    plot(rev(le$G..A[1:25]), type = "b",
         col = myColor_GA, bty = "n", 
          lwd = 1.5, 
         cex = cex, axes = F, xlab = "3' position", ylab = NA, ylim = c(0, y2))
     axis(side = 1, at = seq(1,25,5), labels = seq(25, 1,-5))
     axis(side = 4)
  }
  plots_left <- nCol*2 - length(myLibs)*2
  while(plots_left > 0){
    blank_plot()
    plots_left <- plots_left - 1
  }

}

  legend(1,10, legend = c("C -> T", "G -> A"), col = c(myColor_CT, myColor_GA),
       lty = 1, pch = 21, bty = "n", cex = 2)
# blank_plot()


dev.off()
```

