---
title: "Dstat_genotype_calling_comparison"
author: "Cruz-Dávalos, Diana I."
date: "10/14/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F)
```

```{r}
library(ggplot2)
library(maps)
library(plyr)
source("~/Projects/Botocudos/Scripts/Plotting/plot_F3.R")
source("~/Projects/Botocudos/Scripts/Plotting/plot_Dstats.R")
library(colorspace)
library(cowplot)
panel <- read.table("~/Projects/Botocudos/Files/Panels/Magic.panel", 
                    header = T, stringsAsFactors = F)
magic <- read.table("~/Projects/Botocudos/Files/Panels/Magic.panel", 
                    header = T, stringsAsFactors = F)
languages <- read.table("~/Projects/Botocudos/Files/Panels/Maanasa/Maanasa.americas.Moreno_names.txt", header = F, stringsAsFactors = F)
colnames(languages) <- c("population", "language")
l2 <- data.frame(population = c("Inuit", "USR1", "SpCave", 
                                "Anzick1",  "Lovelock2",
                                "Lovelock3", "LagoaSta", "LagoaSanta_called","Ayayema",
                                "Taino", 
                                "Athabascan",  "Yukpa", "MN0008_L3U"),
                 language = c("Inuit", "AncBeringian", "AncNorth",
                              "AncNorth", "AncNorth", 
                              "AncNorth", "AncSouth", "AncSouth", "AncSouth",
                              "HistSouth", 
                              "Athabascan", "Ge-PanoCaribean", "Botocudo"))

languages <- rbind(languages, l2)
boto <- read.table("~/Projects/Botocudos/Files/Summaries/2019_07_26/Sup_Table_1.txt", header = T, sep = "\t")
```

## Z-scores from qpDstat

```{r}
# switch_h1_h2 <- function(Dstat, pop_in_h1){
#   idx_to_switch <- which(Dstat$h2 == pop_in_h1)
#   h2 <- Dstat$h1[idx_to_switch]
#   Dstat$h1[idx_to_switch] <- pop_in_h1
#   Dstat$h2[idx_to_switch] <- h2
#   Dstat$Z[idx_to_switch] <- -Dstat$Z[idx_to_switch]
#   return(Dstat)
# }
```

```{r}

make_plot_comparison <- function(myComparison, region){
  if(sum(!(region %in% unique(myComparison$region))) > 0){
    myRegion <- "all other regions"
  }else{
    myRegion <- region
  }
  myComparison <- myComparison[myComparison$region %in% region,]
  myCor <- cor(myComparison$victor, myComparison$diana)
  
  main_txt <- paste("Z-scores for D(LagoaSta, NatAm; ",
                    myRegion, ", Yoruba)", sep = "")
  
  plot(myComparison$victor, myComparison$diana,
       xlab = "H1 = LagoaSta, calls by Víctor", 
       ylab = "H1 = LagoaSta, calls by Diana", 
       main = main_txt,
       bty = "n", col = as.factor(myComparison$region))
  grid()
  abline(a = 0, b = 1, col = "red")
  text(paste("Correlation:", round(myCor,4)), 
       x = min(myComparison$victor), y = max(myComparison$diana))
}


subset_dstat <- function(Dstat, pop){
  new_dstat <- Dstat[Dstat$H1 == pop | Dstat$H2 == pop,]
  new_dstat <- switch_h1_h2(Dstat = new_dstat, pop_in_h1 = pop)
  new_dstat$h2_h3 <- paste(new_dstat$H2, new_dstat$H3, sep = "_")
  new_dstat <-new_dstat[new_dstat$H2 != new_dstat$H3,]
  new_dstat <- unique(new_dstat)
  return(new_dstat)
}

```

```{r}

myDirs <- c("MyParams/", "Moreno2019/", "Moreno2019_ab25/")

for(D in myDirs){
  path <- paste("~/Projects/Botocudos/Files/Dstat/2019_10_14/geno_calls/", D,
                "h_merged_qpDstat.results", sep = "")
  Dstat <- read.table(path,
                      stringsAsFactors = F)
  colnames(Dstat) <- c("H1", "H2", "H3", "H4",
                       "D", "Z", "BABA", "ABBA", "nSites")
  Dstat$population <- Dstat$H3
  
  Dstat <- join(Dstat, unique(magic[,c("population", "region")]), by = "population")
  
  
  victor <- subset_dstat(Dstat, "LagoaSta")
  victor <- victor[victor$H2 != "LagoaSanta_called",]
  
  diana <- subset_dstat(Dstat, "LagoaSanta_called")
  diana <- diana[diana$H2 != "LagoaSta",]
  
  myComparison <- diana[,c("h2_h3", "Z", "population", "region")]
  myComparison$diana <- myComparison$Z
  myComparison$Z <- NULL
  
  myComparison <- join(myComparison, victor[,c("h2_h3", "Z")], by = "h2_h3")
  myComparison$victor <- myComparison$Z
  myComparison <- myComparison[complete.cases(myComparison),]
  
  layout(matrix(seq(1,12), ncol = 3))
  for(r in c("Siberia", "EastAsia", "CentralAsia",
             "Caucasus", "Oceania", "SEAsia", "SouthAsia",
             "Europe", "WestAsia", "Africa", "Archaic" )){ #unique(myComparison$region)){
    make_plot_comparison(myComparison, r)
  }
  print(table(myComparison$region))
  print(table(victor$region))
  print(table(diana$region))
  print(sum(table(myComparison$region)))
}
```

### Australasians vs others
```{r}
Dstat <- read.table("~/Projects/Botocudos/Files/Dstat/2019_10_14/h_merged_qpDstat_Oct22.results",
                    stringsAsFactors = F)
colnames(Dstat) <- c("H1", "H2", "H3", "H4",
                     "D", "Z", "BABA", "ABBA", "nSites")
Dstat$population <- Dstat$H3

Dstat <- join(Dstat, magic[,c("population", "region")], by = "population")

```

```{r, fig.width=12, fig.height=12}
column_to_highlight <- "region"
to_highlight <- c("Oceania", "SEAsia")
color_to_highlight <- c("purple", "pink")

pop_in_h3 <- unique(Dstat$H3)
pop_order <- unique(c(Dstat$H1, Dstat$H2))

myPops <- c("MN0008_L3U", "MN0008_3lanes",
            "Surui", "Karitiana", "LagoaSanta_called",
            "LagoaSta",
            "Mixe", "Huichol", "Quechua", "Ayayema")
layout(matrix(seq(1,10), ncol = 5, byrow = T))

for(pop_in_h2 in myPops){
  mySubset <- prepare_set(abba = Dstat, pop_in_h2 = pop_in_h2,
                        pop_in_h3 = pop_in_h3, 
                        pop_order = pop_order)
moreno_4a(abba = mySubset, pop = pop_in_h2,
          column_to_highlight = column_to_highlight, 
          to_highlight = to_highlight, 
          color_to_highlight = color_to_highlight)
}
pop_in_h2 <- "LagoaSanta_called"





```

## Relationship between Botocudos and other Native Americans

```{r, old results}

coordinates <- read.table("~/Projects/Botocudos/Files/Maps/Magic.coordinates", header = F)
colnames(coordinates) <- c("population", "lat", "long", "type")

f3 <- read.table("~/Projects/Botocudos/Files/F3/2019_05_17/boto_maanasa.txt",
                 header = T)
colnames(coordinates) <- c("Source2", "lat", "long", "type")
f3$Source2 <- sub("Conchal.*", "ConchaliSantiagoRM", f3$Source2)
f3 <- join(f3, coordinates, by = "Source2")
f3 <- f3[f3$SNPs >10000,]

f3$population <- f3$Source2

myPanel <- unique(panel[,c("population", "region")])
f3 <- join(f3, myPanel, by = "population")
 f3 <- join(f3, languages, by = "population")
f3 <- f3[f3$region == "Americas",]

```

```{r, fig.height=20, eval = F}
ciempies_plot(f3, main = NULL)
```

# F3
# Arara, Surui, LagoaSanta, Karitiana

```{r}
f3 <- read.table("~/Projects/Botocudos/Files/F3/2019_10_17/f3.results",
                 header = T, stringsAsFactors = F)

f3$Source2 <- sub("Conchal.*", "ConchaliSantiagoRM", f3$Source2)
f3 <- join(f3, coordinates, by = "Source2")
f3 <- f3[f3$SNPs >10000,]

f3$population <- f3$Source2

```

```{r, fig.width=12, fig.height=14}
myPops <- c("Botocudos", "L3U",
            "Arara", "LagoaSanta", "Surui",
             "Karitiana")
# pdf("~/Projects/Botocudos/Plots/F3/2019_10_17/f3_bidir.pdf",
    # width = 12, height = 14)
layout(matrix(seq(1,6), byrow = T, ncol = 6))
for(p in myPops){
  ciempies_plot(f3[f3$Source1 == p,], main = p, 
                doLayout = F, doBar = F, cex.axis = 0.7)
}
# dev.off()
```


```{r}
m <- matrix(c(          "USR1",   "black",    22, "turquoise",
                  "AncBeringian",   "black",    22, "turquoise",
                "NaDene", "#FFE600",    16,      NA,
           "EskimoAleut", "#FFE600",    17,      NA,
              "Saqqaq",   "black",    21,   "#FFE600",
          "TrailCreek",   "black",    23,   "#99B333",
          "SpiritCave",   "black",    21,   "#99B333",
              "Clovis",   "black",    22,   "#99B333",
            "Lovelock",   "black",    24,   "#99B333",
      "NorthernAmerind", "#00D973",    16,      NA,
 "SouthWesternOntario",   "black",    21,   "#00D973",
         "CentralAmerind", "#D94D99",    16,      NA,
       "ChibchanPaezan", "#D94D99",    18,      NA,
              "Belize",   "black",    21,   "#D94D99",
                 "Andean", "#EBD999",    16,      NA,
          "EarlyAndes",   "black",    22,   "#EBD999",
    "LateCentralAndes",   "black",    21,   "#EBD999",
       "Southern_Cone",   "black",    23,   "#EBD999",
     "HistoricalAndean",   "black",    24,   "#EBD999",
    "EcuatorialTucanoan", "#F30000",    16,      NA,
      "Ge-PanoCaribean", "#F30000",    17,      NA,
              "Brazil",   "black",    22,   "#F30000",
          "LagoaSanta",   "black",    21,   "#F30000",
               "Taino",   "black",    23,   "#F30000",
           "Aconcagua",   "black",    21,   "#6B9346",
       "PuntaSantaAna",   "black",    22,   "#6B9346",
       "AncientYamana",   "black",    23,   "#6B9346",
             "Ayayema",   "black",    24,   "#6B9346",
     "AncientKaweskar",   "black",    25,   "#6B9346",
              "Clovis",   "black",     16,   "gray",
            "Mexican",    "gray",     16, NA,
       "AncNorth",        "gray20",   16, NA,
       "AncSouth",        "gray80",   16, NA,
       "HistSouth",       "gray60",   16, NA,
      "Other Language",   "gray",     16,    NA,
           "Botocudos",   "black",    16,   "#000D50"), ncol = 4, byrow = T)

myLegend <- data.frame(language = m[,1],
                       bg = m[,2],
                       shape = as.integer(m[,3]),
                       color = m[,4])
myLegend$language <- as.character(myLegend$language)
myLegend$fill <- ifelse(myLegend$bg == "black", as.character(myLegend$color),
                        as.character(myLegend$bg))
#myLegend$shape <- as.integer(myLegend$shape)

```

```{r}
require(RColorBrewer)
myF3 <- reshape(f3[,c("f_3", "Source1", "Source2")], direction = "wide", 
        timevar  = "Source2",  idvar = c("Source1"))
rownames(myF3) <- myF3$Source1
myF3$Source1 <- NULL
myF3 <-  data.matrix(myF3, rownames.force = T)

myF3[is.na(myF3)] <- 0.3
coul <- colorRampPalette(brewer.pal(8, "PiYG"))(5)
heatmap(myF3, col = magma(50), keep.dendro = F)

```

```{r}
require(forcats)
require(dplyr)


f3_to_plot <- f3 %>% filter(Source1 == "L3U" & Source2 != "Botocudos" & Source2 != "Mexican")
f3_to_plot$Source2 <- factor(f3_to_plot$Source2) 
f3_to_plot$Source2 <- fct_reorder(f3_to_plot$Source2, f3_to_plot$f_3, .desc = T)
f3_to_plot <- f3_to_plot %>% left_join(languages, "population") %>% left_join(myLegend, "language") %>%
 mutate(language = ifelse(is.na(language),  "Other Language", language))
f3_to_plot <- f3_to_plot[order(f3_to_plot$f_3, decreasing = T),]

theLanguages <- c("Ge-PanoCaribean", "EcuatorialTucanoan",
                  "ChibchanPaezan", "CentralAmerind", 
                  "Andean", "NorthernAmerind", "Clovis", "AncSouth",
                  "HistSouth", "HistoricalAndean", "NaDene", "EskimoAleut",
                  "AncNorth",
                  "AncBeringian", "Saqqaq", "Other Language")

for(l in theLanguages){

  png(paste("~/Projects/Botocudos/Plots/F3/2019_10_17/f3_", l, ".png", sep = ""),
      width = 14, height = 5, res = 250, units = "in", bg = NA)
      plot(1:nrow(f3_lang), ylim = c(0.2, 0.3), type = "n", axes = F, xlab = NA, ylab = NA)
    grid()
  i <- 1
  myLang <- theLanguages[i]
  while(l != myLang){
    i <- i+1
    f3_lang <- f3_to_plot %>% 
      mutate(shape = case_when(language == myLang ~ shape,
                               T ~ NA_integer_)) %>%
      mutate(fill = case_when(language == myLang ~ fill,
                              T ~ NA_character_)) %>%
      mutate(std.err = case_when(language == myLang ~ std.err,
                                 T ~ NA_real_))
    
    # layout(matrix(seq(1,2), ncol = 1), widths = c(1), heights = c(2,5))
    

    segments(x0 = 1:nrow(f3_lang), y0 = f3_lang$f_3 - qnorm(0.995)*f3_lang$std.err,
             y1 = f3_lang$f_3 + qnorm(0.995)*f3_lang$std.err )
    points(x = 1:length(f3_lang$Source2), y = f3_lang$f_3, 
           pch = f3_lang$shape,
           col = as.character(f3_lang$fill),  bg = as.character(f3_lang$fill), cex = 1.2
    )
    myLang <- theLanguages[i]
  }
    axis(2)
    axis(1, at = 1:nrow(f3_lang), labels = f3_lang$Source2, las = 2, cex.axis = 0.6)
    dev.off()
}
```


```{r}
for(l in theLanguages){

  png(paste("~/Projects/Botocudos/Plots/F3/2019_10_17/f3_map_", l, ".png", sep = ""),
      width = 4, height = 4, res = 250, units = "in", bg = NA)
          par(fg = NULL)
    map('world', fill = T, col = "gray80",
        xlim=c(-168,-10), ylim = c(-60, 100),
        border="white", lforce = "e")
  i <- 1
  myLang <- theLanguages[i]
  while(l != myLang){
        i <- i+1
    f3_lang <- f3_to_plot %>% 
      mutate(shape = case_when(language == myLang ~ shape,
                               T ~ NA_integer_)) %>%
      mutate(fill = case_when(language == myLang ~ fill,
                              T ~ NA_character_)) %>%
      mutate(std.err = case_when(language == myLang ~ std.err,
                                 T ~ NA_real_))
    points(x = f3_lang$long, y = f3_lang$lat, pch = f3_lang$shape,
           col = f3_lang$fill, bg = f3_lang$color, cex = 1.5)
    
    myLang <- theLanguages[i]
  }
  dev.off()
}
```

```{r}



plot(1:10, 1:10, axes = F, xlab=NA, ylab=NA, type = "n", bty = "n")
legend(x = 2, y = 9, bty = "n", ncol = 1, legend = myLegend$language,
       pt.bg = myLegend$bg,
       pch = myLegend$shape,
       col = myLegend$color, pt.lwd = 1.2,  cex = 0.5)

plot(1:nrow(f3_to_plot), ylim = c(0.2, 0.3), type = "n", axes = F, xlab = NA, ylab = NA)
grid()
segments(x0 = 1:nrow(f3_to_plot), y0 = f3_to_plot$f_3 - qnorm(0.995)*f3_to_plot$std.err,
          y1 = f3_to_plot$f_3 + qnorm(0.995)*f3_to_plot$std.err )
points(x = 1:length(f3_to_plot$Source2), y = f3_to_plot$f_3, 
     pch = f3_to_plot$shape,
     col = as.character(f3_to_plot$fill),  bg = as.character(f3_to_plot$fill), cex = 2
     )

```


```{r}
ggplot(aes( y = f_3, x = Source2, 
              ymin = f_3 - qnorm(0.995)*std.err,
              ymax = f_3 + qnorm(0.995)*std.err, 
              color = language)) +
  geom_point() +
  geom_errorbar(lwd = 2) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        panel.grid.major = element_line(size = 0.25, linetype = 'dashed',
                                colour = "gray")) +
  labs(x = NULL, y = NULL) +
  scale_color_manual(breaks = myLegend$language,
                     values = myLegend$fill,
                     labels = myLegend$language)



```
```{r f3 for best hits}
myPop <- "Arara"
f3_to_plot <- f3 %>% filter(Source1 == myPop &
                              Source2 != "Botocudos" & 
                              Source2 != "Mexican") %>%
  mutate(Source2 = case_when(Source2 == "L3U" ~ "Botocudos",
                             T ~ Source2)) 

f3_to_plot$Source2 <- factor(f3_to_plot$Source2) 
f3_to_plot$Source2 <- fct_reorder(f3_to_plot$Source2,
                                  f3_to_plot$f_3,
                                  .desc = T)

f3_to_plot <- f3_to_plot %>% left_join(languages, "population") %>% left_join(myLegend, "language") %>%
  mutate(language = ifelse(is.na(language),  
                           "Other Language", language)) %>%
  mutate(fill = case_when(Source2 == "Botocudos" ~ "black",
                          T                      ~ fill)) %>%
  mutate(shape = case_when(Source2 == "Botocudos" ~ 16.0,
                           T                      ~ shape)) %>%
  mutate(bg = case_when(Source2 == "Botocudos" ~ "black",
                        T                      ~ bg))
f3_to_plot <- f3_to_plot[order(f3_to_plot$f_3, decreasing = T),]


plot(1:nrow(f3_to_plot), ylim = c(0.2, 0.3), type = "n", axes = F, xlab = NA, ylab = NA)
grid()
segments(x0 = 1:nrow(f3_to_plot), y0 = f3_to_plot$f_3 - qnorm(0.995)*f3_to_plot$std.err,
          y1 = f3_to_plot$f_3 + qnorm(0.995)*f3_to_plot$std.err )
points(x = 1:length(f3_to_plot$Source2), y = f3_to_plot$f_3, 
     pch = f3_to_plot$shape,
     col = as.character(f3_to_plot$fill),  bg = as.character(f3_to_plot$fill), cex = 2
     )
axis(2)
    axis(1, at = 1:nrow(f3_lang), labels = f3_lang$Source2, las = 2, cex.axis = 0.6)
```

# ANGSD, error corrected
  # do not trim 5bp, 
  # MN0008
  # All botocudos
  # no need to rmTrans
  
  # Fig. 3C
questions is:
Compared to other Native Americans, how close are botocudos to Mixe or to Lagoa Santa?


```{r}
nat_americans <- list("northern" = c("Lovelock2", "Lovelock3",
                                     "Pima", "Huichol", "Maya", "Taino"),
                      "east_andes" = c("Yukpa", "Aconcagua",
                                       "Aymara", "Quechua", 
                                       "Chane", "Surui", "Karitiana",
                                       "Piapoco"),
                      "west_andes" = c("Ayayema", "PtaStaAna",
                                       "AncKaweskar2", "AncKaweskar3",
                                       "AncYamana", "AncYamana2"),
                      "botocudos" = c("Botocudos"))
```


```{r}

```


### rmTrans
```{r, fig.width=12, fig.height=6}
abba <- read.table("~/Projects/Botocudos/Files/Dstat/2019_10_14/Corrected/23Boto_inc.ErrorCorr.TransRem.txt",
                   header = T, stringsAsFactors = F)

# first plot is D(NatAm, Mixe; LagoaSanta_15x, Yoruba)
myAlpha <- 0.001
pop_order <- rev(c("Botocudos", "MN0008_L3U",
               "Surui", "Karitiana", "Ayayema", "Aymara","Huichol"))



layout(matrix(seq(1,2), byrow = T, ncol = 2))
first_set <- prepare_set(abba, pop_in_h2 = "Mixe", 
                         pop_in_h3 = "LagoaSanta_15x",
                         pop_order = pop_order)


cex <- 1.2
lwd <- 2
par(mar = c(5,8,4,2))
moreno_3c(first_set, 
          title =  "D(NatAm, Mixe; LagoaSanta_15x, Yoruba)\n(rmTrans)",
          lwd = lwd, cex = cex)

# second plot is D(NatAm, LagoaSanta_15X; Mixe, Yoruba)
second_set <- prepare_set(abba, pop_in_h3 = "Mixe", 
                         pop_in_h2 = "LagoaSanta_15x",
                         pop_order = pop_order)

par(mar = c(5,2,4,8))
moreno_3c(second_set, 
          title =  "D(NatAm, LagoaSanta_15x; Mixe, Yoruba)\n(rmTrans)",
          yaxis = F, lwd = lwd, cex = cex)


```

### all
```{r,  fig.width=12, fig.height=6}
abba <- read.table("~/Projects/Botocudos/Files/Dstat/2019_10_14/Corrected/23Boto_inc.ErrorCorr.txt",
                   header = T, stringsAsFactors = F)

# first plot is D(NatAm, Mixe; LagoaSanta_15x, Yoruba)
myAlpha <- 0.001
pop_order <- rev(c("Botocudos", "MN0008_L3U",
               "Surui", "Karitiana", "Ayayema", "Aymara","Huichol"))

cex <- 1.2
lwd <- 2
color <- "#0057ba"
layout(matrix(seq(1,2), byrow = T, ncol = 2))
first_set <- prepare_set(abba, pop_in_h2 = "Mixe", 
                         pop_in_h3 = "LagoaSanta_15x",
                         pop_order = pop_order)

par(mar = c(5,8,4,2))

moreno_3c(first_set, 
          title =  "D(NatAm, Mixe; LagoaSanta_15x, Yoruba)\n(all)",
          xlim = c(-0.7, 0.05), cex = cex, lwd = lwd,
          color = color)

# second plot is D(NatAm, LagoaSanta_15X; Mixe, Yoruba)
second_set <- prepare_set(abba, pop_in_h3 = "Mixe", 
                         pop_in_h2 = "LagoaSanta_15x",
                         pop_order = pop_order)
par(mar = c(5,2,4,8))

moreno_3c(second_set, 
          title =  "D(NatAm, LagoaSanta_15x; Mixe, Yoruba)\n(all)",
           cex = cex, lwd = lwd,
          color = color, yaxis = F)

```


## Figure S35A

### Error-corrected D-stats, rmTrans

```{r, fig.width=12, fig.height=12}
abba <- read.table("~/Projects/Botocudos/Files/Dstat/2019_10_14/Corrected/S35A_Bot8.ErrorCorr.TransRem.txt",
                   header = T, stringsAsFactors = F)

pops_in_title <- c("Botocudos", 
                   "MN0008_L3U", "LagoaSanta_15x",
                   "Surui", "Karitiana",
                   "Mixe", "Huichol", 
                   "Aymara", "Ayayema")

pops_in_y <- pops_in_title
pops_in_h3 <- c("French", "Han",
                          "Andaman", "WCDesert")
pop_colors <- c("deepskyblue1", "darkgoldenrod1", 
                "pink", "blueviolet", 
                "black")

plot_list <- moreno_S35A(abba = abba, 
            pops_in_title = pops_in_title,
            pops_in_y = pops_in_y,
            pops_in_h3 = pops_in_h3,
            pop_colors = pop_colors,
            rm_legend = T)

myPlot <- moreno_S35A(abba = abba,
                      pops_in_title = c("Botocudos"),
                      pops_in_y = pops_in_y,
                      pops_in_h3 = pops_in_h3,
                      pop_colors = pop_colors, rm_legend = F)

legend <- get_legend(myPlot[[1]])
plot_list[[length(plot_list)+1]] <- legend
plot_grid(plotlist = plot_list, ncol = 3)

```

```{r}
require(dplyr)
pop_order <- rev(c("Botocudos",
               "Surui", "Karitiana", "Ayayema", "Aymara","Huichol", "Mixe"))

myAbba <- abba %>% filter(H1 != "Botocudos" & H2 != "Botocudos") %>%
                    mutate(H1 = case_when(H1 == "MN0008_L3U" ~ "Botocudos",
                                          T ~ H1)) %>%
                    mutate(H2 = case_when(H2 == "MN0008_L3U" ~ "Botocudos",
                                          T ~ H2)) %>%
  filter(H1 %in% pop_order & H2 %in% pop_order); unique(myAbba$H1) ; unique(myAbba$H2)


```

```{r}
myAlpha <- 0.001

png("~/Projects/Botocudos/Plots/Dstat/dbc_2019_dic.png",
    bg = NA, width = 12, height = 4, res = 300, units = "in")
cex <- 1.5
lwd <- 2
color <- "#0057ba"
layout(matrix(seq(1,2), byrow = T, ncol = 2))
first_set <- prepare_set(myAbba, pop_in_h2 = "Mixe", 
                         pop_in_h3 = "WCDesert",
                         pop_order = pop_order)

# second plot is D(NatAm, LagoaSanta_15X; Mixe, Yoruba)
par(mar = c(5,8,4,2))

moreno_3c(first_set, 
          title =  "D(NatAm, Mixe; Australian, Yoruba)",
          xlim = c(-0.05, 0.05), cex = cex, lwd = lwd,
          color = color, annotate = F)

second_set <- prepare_set(myAbba, pop_in_h3 = "Andaman", 
                         pop_in_h2 = "Mixe",
                         pop_order = pop_order)
par(mar = c(5,2,4,8))
moreno_3c(second_set, 
          title =  "D(NatAm, Mixe; Andaman, Yoruba)",
           cex = cex, lwd = lwd,
          color = color, yaxis = F, annotate = F)

dev.off()
```


#### Affinities of Native Americans to non-NA (vs Mixe) 
```{r, fig.width=12, fig.height=12}

pops_to_plot <- c("WCDesert", "Andaman","Han", "French")
colors <- c("blueviolet", "pink", 
             "darkgoldenrod1", "deepskyblue1", 
                "black")

layout(matrix(seq(1,4), byrow = T, ncol = 2))
for(i in 1:4){
  pop <- pops_to_plot[i]
  color <- colors[i]
  myAbba <- prepare_set(abba, pop_in_h2 = "Mixe", pop_in_h3 = pop, pop_order = c("Botocudos", 
                                                                                        "MN0008_L3U", "LagoaSanta_15x",
                                                                                        "Surui", "Karitiana",
                                                                                        "Mixe", "Huichol", 
                                                                                        "Aymara", "Ayayema"))
  
  myAbba <- myAbba[complete.cases(myAbba),]
  
  moreno_3c(myAbba, 
            title = paste("D(NatAm, Mixe; ", pop, ", Yoruba)"),
            color = color)
}


```

#### Affinities of Native Americans to non-NA (vs Lagoa Santa)

```{r, fig.width=12, fig.height=12}
pops_to_plot <- c("WCDesert", "Andaman","Han", "French")
colors <- c("blueviolet", "pink", 
             "darkgoldenrod1", "deepskyblue1", 
                "black")

layout(matrix(seq(1,4), byrow = T, ncol = 2))
for(i in 1:4){
  pop <- pops_to_plot[i]
  color <- colors[i]
  myAbba <- prepare_set(abba, pop_in_h2 = "LagoaSanta_15x", pop_in_h3 = pop, pop_order = c("Botocudos", 
                                                                                        "MN0008_L3U", "LagoaSanta_15x",
                                                                                        "Surui", "Karitiana",
                                                                                        "Mixe", "Huichol", 
                                                                                        "Aymara", "Ayayema"))
  
  myAbba <- myAbba[complete.cases(myAbba),]
  
  moreno_3c(myAbba, 
            title = paste("D(NatAm, LagoaSta; ", pop, ", Yoruba)"),
            color = color)
}


```

## Error-corrected D-stats, all type of polymorphisms
```{r, fig.width=12, fig.height=12}
abba <- read.table("~/Projects/Botocudos/Files/Dstat/2019_10_14/Corrected/S35A_Bot8.ErrorCorr.txt",
                   header = T, stringsAsFactors = F)

plot_list <- moreno_S35A(abba = abba, 
            pops_in_title = pops_in_title,
            pops_in_y = pops_in_title,
            pops_in_h3 = pops_in_h3,
            pop_colors = pop_colors,
            rm_legend = T)
myPlot <- moreno_S35A(abba = abba,
                      pops_in_title = c("Botocudos"),
                      pops_in_y = pops_in_y,
                      pops_in_h3 = pops_in_h3,
                      pop_colors = pop_colors, rm_legend = F)

legend <- get_legend(myPlot[[1]])
plot_list[[length(plot_list)+1]] <- legend
plot_grid(plotlist = plot_list, ncol = 3)

```


# D-stats, called genotypes
```{r}

```

# Enhanced D-stats
## Neanderthal
```{r}
Dstat <- read.table("~/Projects/Botocudos/Files/Dstat/2019_10_14/Enhanced/nean_h_merged_qpDstat.results",
                    stringsAsFactors = F)
colnames(Dstat) <- c("H1", "H2", "H3", "H4",
                     "D", "Z", "BABA", "ABBA", "nSites")
Dstat$population <- Dstat$H3

Dstat <- join(Dstat, magic[,c("population", "region")], by = "population")


```


```{r}
column_to_highlight <- "population"
to_highlight <- c("AltaiNean")
color_to_highlight <- c("seagreen2")

pop_in_h3 <- unique(Dstat$H3)
pop_order <- unique(c(Dstat$H1, Dstat$H2))

myLanguages <- (c("Botocudo", "Ge-PanoCaribean", "HistSouth",
                     "Andean", "AncSouth", "EcuatorialTucanoan", 
                     "NorthernAmerind", "CentralAmerind", "Athabascan",
                     "Inuit", "Saqqaq", "AncNorth", "AncBeringian")) #unique(mySubset$language)
groups_h1 <- myLanguages
tick_color <- c("black", "peachpuff4", "peachpuff3", 
                "palegreen3", "paleturquoise3", "paleturquoise4", 
                "palevioletred", "palevioletred4", "royalblue2", 
                "red3", "red4", "salmon3", "sandybrown")

myPops_H1 <- rev(c("MN0008_L3U", "Yukpa","Karitiana","Taino", 
                    "Aymara",
                   "LagoaSanta_called",
                   "LagoaSta", "Piapoco", "Maya",
                   "Mixe", "Huichol", "Pima",
                   "Athabascan","Surui",
                   "Chane",
                   "Quechua",
                   "Inuit", "Saqqaq",
                   "SpCave", "Anzick1", "USR1", 
                   "Ayayema","Lovelock3", "Lovelock2"))
png("~/Projects/Botocudos/Plots/Dstat/Enhanced_neanderthal.png",
    width = 2.5*ncol, height = 4*ncol, res = 250, units = "in")

ncol <- round(sqrt(length(unique(c(Dstat$H1, Dstat$H2)))))
layout(matrix(seq(1, ncol**2), ncol = ncol, byrow = T))

for(pop_in_h2 in myPops_H1){
  mySubset <- prepare_set(abba = Dstat, pop_in_h2 = pop_in_h2,
                          pop_in_h3 = pop_in_h3, 
                          pop_order = pop_order)
  languages$H1 <- languages$population
  mySubset <- join(mySubset, languages[,c("H1", "language")], by = "H1")
  
  order_h1 <- myPops_H1
  moreno_4a(abba = mySubset, pop = pop_in_h2,
            column_to_highlight = column_to_highlight, 
            to_highlight = to_highlight, 
            color_to_highlight = color_to_highlight, order_h1 = myPops_H1,
            cex = 1.5, lwd = 2.5, x1 = -6, x2 = 6, tick_color = tick_color,
            groups_h1 = groups_h1, col_groups_h1 = "language")
}
moreno_4a(abba = mySubset, pop = pop_in_h2,
          column_to_highlight = column_to_highlight, 
          to_highlight = to_highlight, 
          color_to_highlight = color_to_highlight, order_h1 = myPops_H1,
          cex = 2, lwd = 2.5, plot = F, legend = T,
          tick_color = tick_color,
            groups_h1 = groups_h1, col_groups_h1 = "language")
dev.off()
```

```{r}
abba <- Dstat 
# first plot is D(NatAm, Lovelock2; AltaiNean, Yoruba)
myAlpha <- 0.001
pop_order <- myPops_H1
pop_in_h2 <- "Lovelock2"
pop_in_h3 <- "AltaiNean"
first_set <- prepare_set(abba, pop_in_h2 = pop_in_h2, 
                         pop_in_h3 = pop_in_h3,
                         pop_order = pop_order)

png(paste("~/Projects/Botocudos/Plots/Dstat/enhanced_Dstat_h2=", pop_in_h2,
          "_h3=", pop_in_h3,".png",
          sep = ""),
    width = 7, height = 7, res = 250, units = "in")
cex <- 1.2
lwd <- 2
par(mar = c(5,8,4,2))
moreno_3c(first_set, 
          title =  paste("D(NatAm, ", pop_in_h2, "; ",pop_in_h3,", Yoruba)", sep = ""),
          lwd = lwd, cex = cex, col_to_plot = "D", xlim = c(-.5,.5))
dev.off()

```

## Denisova
```{r}
Dstat <- read.table("~/Projects/Botocudos/Files/Dstat/2019_10_14/Enhanced/deni_h_merged_qpDstat.results",
                    stringsAsFactors = F)
colnames(Dstat) <- c("H1", "H2", "H3", "H4",
                     "D", "Z", "BABA", "ABBA", "nSites")
Dstat$population <- Dstat$H3

Dstat <- join(Dstat, magic[,c("population", "region")], by = "population")
```


```{r}
column_to_highlight <- "population"
to_highlight <- c("Denisova")
color_to_highlight <- c("coral")

pop_in_h3 <- unique(Dstat$H3)
pop_order <- unique(c(Dstat$H1, Dstat$H2))

myPops_H1 <- rev(c("MN0008_L3U", "Taino", "Karitiana",
                   "LagoaSanta_called",
                   "LagoaSta","Mixe","Aymara", "Piapoco", 
                   "Yukpa","Surui","Maya",
                   "SpCave",   
                    "Huichol", 
                   "Athabascan",
                   "Chane","Pima",
                   "Quechua",
                   "Inuit", "Saqqaq",
                   "Ayayema","Lovelock3", "Lovelock2", "Anzick1","USR1"))


png("~/Projects/Botocudos/Plots/Dstat/Enhanced_denisova.png",
    width = 2.5*ncol, height = 4*ncol, res = 250, units = "in")

ncol <- round(sqrt(length(unique(c(Dstat$H1, Dstat$H2)))))
layout(matrix(seq(1, ncol**2), ncol = ncol, byrow = T))

for(pop_in_h2 in myPops_H1){
  mySubset <- prepare_set(abba = Dstat, pop_in_h2 = pop_in_h2,
                        pop_in_h3 = pop_in_h3, 
                        pop_order = pop_order)
moreno_4a(abba = mySubset, pop = pop_in_h2,
          column_to_highlight = column_to_highlight, 
          to_highlight = to_highlight, 
          color_to_highlight = color_to_highlight,
          order_h1 = myPops_H1, cex = 1.5, x1 = -4, x2 = 4, lwd = 2.5)
}
moreno_4a(abba = mySubset, pop = pop_in_h2,
          column_to_highlight = column_to_highlight, 
          to_highlight = to_highlight, 
          color_to_highlight = color_to_highlight, order_h1 = myPops_H1,
          cex = 2, lwd = 2.5, plot = F, legend = T)
dev.off()
```


```{r}
abba <- Dstat 
# first plot is D(NatAm, Lovelock2; AltaiNean, Yoruba)
myAlpha <- 0.001
pop_order <- myPops_H1

first_set <- prepare_set(abba, pop_in_h2 = "Anzick1", 
                         pop_in_h3 = "Denisova",
                         pop_order = pop_order)


cex <- 1.2
lwd <- 2
par(mar = c(5,8,4,2))
moreno_3c(first_set, 
          title =  "D(NatAm, Lovelock2; Denisova, Yoruba)",
          lwd = lwd, cex = cex, col_to_plot = "D", xlim = c(-.5,.5))

```



# qpGraph
reporta modelo más sencillo, sin admixtures
de todos modos intenta todas las combinaciones

fit score es usado como log-likelihood

fija los admixtures proportions para todas las ramas, excepto para las que estás preguntando

# Admixture graph
# 2012 Reich
# India - qpWave or qpGraph
# Heterozygosity

```{r}
het_dir <- "~/Projects/Botocudos/Files/Heterozygosity/2019_10_30/"

files <- list.files(het_dir, pattern = "Nigeria") ; files

get_samples <- function(pattern){
  x <- sapply(strsplit(files[grep(pattern, files)], "_"), 
              function(x) c(x[[1]], x[[2]]))
  
  samples <- unique(c(x[1,], x[2,]))
  return(samples)
}


boto_sm <- get_samples(pattern = "MN")
boto_sm <-  boto_sm[!boto_sm %in% c("MN1943", "MN01701")]
wcd_sm <- get_samples("WCD")
surui_sm <- get_samples("Surui")
lovelock_sm <- get_samples("Lovelock")
mixe_sm <- get_samples("Mixe")
maya_sm <-get_samples("Maya")
pima_sm <- get_samples("Pima")
usr_sm <- get_samples("USR")

mbuti_sm <- c("HGDP00456", "SS6004471")
dai_sm <- c("SS6004467", "HGDP01307")
french_sm <- c("SS6004468", "HGDP00521")
han_sm <- c("SS6004469", "HGDP00778")
mandenka_sm <- c("SS6004470", "HGDP01284")
papuan_sm <- c("SS6004472", "HGDP00542")
san_sm <- c("SS6004473", "HGDP01029")
sardinian_sm <- c("SS6004474", "HGDP00665")
yoruba_sm <- c("SS6004475", "HGDP00927")
karitiana_sm <- c("SS6004476", "HGDP00998")
australian_sm <- c("SS6004477", "SS6004478")
lagoa_sm <- c("LagoaSta", "LS4", "LS6", "LS7", "LS8")

myGroups <- list(
                 WCD = wcd_sm,
                 Surui = surui_sm,
                 Lovelock = lovelock_sm,
                 Mixe = mixe_sm,
                 Maya = maya_sm,
                 Pima = pima_sm,
                 USR = usr_sm,
                 Mbuti = mbuti_sm,
                 Dai = dai_sm,
                 French = french_sm,
                 Han = han_sm,
                 Mandenka = mandenka_sm,
                 Papuan = papuan_sm,
                 San = san_sm,
                 Sardinian = sardinian_sm,
                 Yoruba = yoruba_sm,
                 Karitiana = karitiana_sm,
                 Australian = australian_sm,
                 LagoaSta = lagoa_sm,
                 Botocudos = boto_sm
                 )
myRegions <- matrix(data = c("WCD", "Australian_WCD", "Oceania",
                             "Surui", "Surui", "Americas",
                             "Lovelock", "Lovelock", "Americas_anc",
                             "Mixe", "Mixe", "Americas",
                             "Maya", "Maya", "Americas",
                             "Pima", "Pima", "Americas",
                             "USR", "USR", "Americas_anc",
                             "Mbuti", "Mbuti", "Africa",
                             "Dai", "Dai", "SouthAsia",
                             "French", "French", "Europe",
                             "Han", "Han", "EastAsia",
                             "Mandenka", "Mandenka", "Africa",
                             "Papuan", "Papuan", "Oceania",
                             "San", "San", "Africa",
                             "Sardinian", "Sardinian", "Europe",
                             "Yoruba", "Yoruba", "Africa",
                             "Karitiana", "Karitiana", "Americas",
                             "Australian", "Australian", "Oceania",
                             "LagoaSta", "LagoaSanta", "Americas_anc",
                             "Botocudos", "Botocudos", "Botocudos"
                             ), ncol = 3, byrow = T)

myRegions <- data.frame(myRegions)
colnames(myRegions) <- c("group", "population", "region")
```

```{r}
match_files <- function(group, files){
  x <- sapply(strsplit(files, "_"), 
              function(x) x[[1]])

  index <- which(x %in% group)
  return(index)
}

lagoa_file <- match_files(lagoa_sm, files = files)


num_files <- function(files){
  myTable <- table(sapply(files, function(x) strsplit(x, "_")[[1]][1]))
  return(myTable)
}

blank_plot <- function(){
  plot(1,1, type = "n", axes = F, xlab = NA, ylab = NA)
}

plot_pairwise_het <- function(samples){
  index <- match_files(samples, files = files)
  
  sub_files <- files[index]
  nFiles <- num_files(sub_files)
  nFiles <- nFiles[order(nFiles, decreasing = T)]
  individuals <- names(nFiles)[order(nFiles, decreasing = T)]
  individuals2 <- c(individuals, samples[!samples %in% individuals])
  nInd <- length(individuals)+1
  
  layout(matrix(1:((nInd-1)*(nInd-1)), byrow = T, ncol = nInd-1))
  
  for(ind1 in individuals){
    individuals2 <- individuals2[!( individuals2 == ind1) ]
    for(ind2 in individuals2){
      to_read <- paste(het_dir, ind1, 
                       "_", ind2, "_Nigeria_B_Yoruba-3.pi.stats.txt",
                       sep = "")
      if(file.exists(to_read)){
        print(to_read)
        myPi <- read.table(to_read, header = F)
        colnames(myPi) <- c("estimate",	"bias",	"stderror",	"confint1",	"confint2")
        
        plot(myPi$estimate, ylim = c(0, 0.5), axes = F,
             xlab = ind1, ylab = ind2)
        segments(x0 = 1, y0 = myPi$confint1, y1 = myPi$confint2)
        axis(2)
        grid()
      }
      else{
        blank_plot()
      }
    }
    if(ind1 != tail(individuals,1)){
      to_plot <- nInd - length(individuals2) -1
      while(to_plot > 0){
        to_plot <- to_plot -1
        blank_plot()
      }
    }
    
  }
}
```

```{r}
plot_pairwise_het(karitiana_sm)

for(group in names(myGroups)){
  nInd <- length(myGroups[[group]])
  png(paste("~/Projects/Botocudos/Plots/Het/2019_11_01/", group, ".pi.pairwise.png", sep = ""),
      width = 2*nInd, height = 2*nInd, re = 250, units = "in")
  plot_pairwise_het(myGroups[[group]])
  dev.off()
}

```

```{r}
myFile <- files[1]
myPi <- read.table(paste(het_dir, myFile, sep = ""), header = F)

ind1 <- strsplit(myFile, "_")[[1]][1]
ind2 <- strsplit(myFile, "_")[[1]][2]


```

```{r}

match_groups <- function(file, groups, regions){
  ind <- strsplit(file, "_")[[1]][1]
  for(group in names(groups)){
    if(ind %in% groups[[group]]){
      return(regions[regions$group == group,])
    }
  }
}

x <- match_groups(files[1], myGroups, myRegions)

```

```{r}
myPi <- data.frame()
for(f in files){
  x <- read.table(paste(het_dir, f, sep = ""), stringsAsFactors = F)

  regions <- match_groups(f, myGroups, myRegions)
  colnames(x) <- c("estimate",	"bias",	"stderror",	"confint1",	"confint2")
  
  # add info about coverage
  x$coverage <- 4
  ind1 <- strsplit(f, "_")[[1]][1]
  ind2 <- strsplit(f, "_")[[1]][2]
  
  if(ind1 %in% boto_sm){
    cov1 <- boto$DoC_endogenous[boto$Library == "All" & boto$Sample == ind1]
    cov2 <- boto$DoC_endogenous[boto$Library == "All" & boto$Sample == ind2]
    x$coverage <- ifelse(cov1*cov2 >4, 4, cov1*cov2)
  }
  x <- cbind(x, regions)
  x$ind1 <- strsplit(f, "_")[[1]][1]
  x$ind2 <- strsplit(f, "_")[[1]][2]
  myPi <- rbind(myPi, x)
}
myPi$region <- factor(myPi$region, levels = c("Botocudos", "Americas", 
                                              "Americas_anc", "Oceania",
                                              "EastAsia", "SouthAsia", "Europe",
                                              "Africa"),
                      ordered = T)
myPi <- myPi[order(myPi$region, myPi$estimate),]

myPi$population <- factor(myPi$population, levels = c("Yoruba", "Mandenka", "Mbuti", "San",
                                                      "French", "Sardinian",
                                                      "Dai",
                                                      "Han",
                                                      "Australian", 
                                                      "Australian_WCD","Papuan",
                                                      "USR", "Lovelock", "LagoaSanta",
                                                      "Maya", "Mixe", "Pima", "Karitiana", "Surui", 
                                                      "Botocudos"),
                          ordered = T)
```

```{r}
pd <- position_dodge(width = 0.4)
```


```{r}
ggplot(myPi, aes(x = population, y = estimate, col = region, alpha = coverage,
                 ymin = confint1, ymax = confint2)) +
  #geom_point() +
  geom_jitter() +
  # geom_errorbar(position = pd) +
  geom_violin() +
  theme(panel.grid.major = element_line(size = 0.5, linetype = 'dashed',
                                colour = "gray"),
         panel.grid.minor = element_line(size = 0.25, linetype = 'dashed',
                                colour = "gray")) +
  scale_color_discrete() +
  scale_alpha_continuous()

```

```{r}
require(dplyr)

to_plot <- myPi %>%
  filter((ifelse(group == "Botocudos", ind2 == "MN0008" & ind1 == "MN00056", T) )) %>% 
  filter(ifelse(group == "LagoaSta", ind1 == "LagoaSta" | ind2 == "LagoaSta", T)) %>%
  filter(ifelse(group == "Pima", ifelse(ind1 == "PimaCEPH", ind2 != "PimaF", T), T)) %>%
  group_by(group) %>%
  sample_n(1)


m <- matrix(ncol = 3, byrow = T,
            c("Botocudos", "black", "Botocudos",
              "Americas", "#bf36e0", "Americas (modern)",
              "Americas_anc", "orange", "Americas (ancient)",
              "Oceania", "#2619d1", "Oceania",
              "EastAsia", "#ffe600", "East Asia",
              "SouthAsia", "lightblue", "SouthAsia",
              "Europe", "#6ea900", "Europe",
              "Africa", "#1b8e13", "Africa"))
myPi_aes <- data.frame(region = m[,1],
                       color = m[,2],
                       label = m[,3])
```


```{r}

png("~/Projects/Botocudos/Plots/Het/2019_11_01/condHet.png", width = 16,
    height = 10,
    res = 300, units = "in", bg = NA)
ggplot(to_plot,
       aes(x = population, y = estimate, col = region, 
                 ymin = confint1, ymax = confint2)) +
  geom_point(size = 2) +
  geom_errorbar(position = pd, width = 0.4, lwd = 1.2) +
  theme(panel.grid.major = element_line(size = 0.5, linetype = 'dashed',
                                colour = "gray"),
         panel.grid.minor = element_line(size = 0.25, linetype = 'dashed',
                                colour = "gray"),
        axis.text.x = element_text(angle = 90, size = 14),
        axis.text.y = element_text(size = 16),
        plot.title = element_text(size=20)) +
  labs(x = NULL, y = NULL, title = "Conditional heterozygosity") +
  scale_color_manual(breaks = myPi_aes$region,
                       values = as.character(myPi_aes$color),
                       label = myPi_aes$label,
                     name = "") +
  scale_alpha_continuous() 
dev.off()

```

```{r}
b <- myPi %>%
  filter(group == "Botocudos")
b2 <- b %>% rename(ind1 = ind2, ind2 = ind1)
  
  ggplot(rbind(b,b2), aes(x = ind1, y = estimate, ymin = confint1, ymax = confint2)) +
  geom_point() +
  geom_errorbar() +
  facet_wrap(ind2 ~ ., ncol =5) +
  theme(axis.text.x =  element_text(angle = 90))


```


### Runs of homozygosity

```{r}

read_roh <- function(file_dir, ind, chromosomes = seq(1,22), snp_type = "all"){
  roh <- data.frame()
  for(chr in chromosomes){
    f <- read.table(paste(file_dir, ind, "_", chr, "_", snp_type, "_roh.hom", sep = ""), 
                    header = T)
    roh <- rbind(roh, f)
  }
  return(roh)
}


plot_roh_ind <- function(ind, file_dir, plot_dir, header){
  
  roh <- read_roh(file_dir, ind)
  header$chr <- sapply(header$V2, function(x) strsplit(x, ":")[[1]][2])
  header$size <- sapply(header$V3, function(x) strsplit(x, ":")[[1]][2])
  
  fig_name <- paste(plot_dir, "ROH_",ind,".png", sep = "")
  png(fig_name,
      height = 7, width = 12, res = 250, units = "in")
  layout(matrix(seq(1, 24), byrow = T, nrow = 4, ncol = 6),
         widths = rep(2,6), heights = rep(1,4))
  
  for(chr in 1:22){
    
    size <- as.integer(header$size[header$chr == chr])
    
    plot(bty = "n",
         ylim = c(0,1), xlim = c(0,3e8),
         type = "n", x = 0, y = 0, xlab = "Coordinates (bp)", ylab = NA,
         main = chr, axes = F)
    axis(1)
    rect(xleft = 1, xright = size,
         ybottom = 0, ytop = 1, lwd = .2,
         col = "gray90")
    rect(xleft = roh$POS1[roh$CHR == chr], xright =roh$POS2[roh$CHR == chr],
         ybottom = 0, ytop = 1, lwd = .2,
         col = "darkcyan") 
    
  }
  dev.off()
}

# chromosomes <- seq(1,22)
# myPath <- "~/Projects/Botocudos/Files/Heterozygosity/2019_10_30/ROH/MN0008_"



```


```{r}
ind <- "MN0008"
file_dir <- "~/Projects/Botocudos/Files/Heterozygosity/2019_10_30/ROH/"
plot_dir <- "~/Projects/Botocudos/Plots/Het/2019_11_01/"
header <- read.table("~/Projects/Botocudos/Files/Heterozygosity/2019_10_30/ROH/MN0008.header", skip = 1, nrows = 22, stringsAsFactors = F)
plot_roh_ind(ind, file_dir, plot_dir, header)
```


```{r}
hist(roh$KB, breaks = 1000, xlim = c(0,2500))

```

```{r}
inds_het <- c("Karitiana",
              "KaritianaB",
              "KaritianaBI16",
              "KaritianaG",
              "KaritianaH",
              "MN0008",
              "SuruiA",
              "SuruiB")

roh_inds <- list(numeric(length(inds_het)))

for(i in 1:length(inds_het)){
  roh_inds[[i]] <- read_roh(file_dir, inds_het[i])
}

names(roh_inds) <- inds_het
max_x <-  sort(sapply(inds_het, 
                      function(x) max(roh_inds[[x]]$KB)), decreasing = T)[1]/1000

nCols <- round(sqrt(length(inds_het)))

layout(matrix(1:(nCols**2), ncol = nCols, byrow = T))
for(ind in inds_het){
  hist(roh_inds[[ind]]$KB/1000, main = ind, 
       xlim = c(0, max_x), 
       ylim = c(0, 300),
       breaks = 100, xlab = "ROH length (Mb)")
}



```

```{r}
ensemble <- read.csv("~/Projects/Botocudos/Files/Heterozygosity/Ensembl/mart_export.txt.gz", sep = "\t")

```

```{r}
roh <- read_roh(file_dir, "MN0008")

get_gene <- function(line){
  gene <- ensemble[(ensemble$Gene.start..bp. < line[7] & ensemble$Gene.end..bp. > line[7]) | 
             (ensemble$Gene.start..bp. < line[8] & ensemble$Gene.end..bp. > line[8]),]
  return(gene)
  
}

genes <- apply(roh, 1, function(x) get_gene(x))

```

